Add KEP for DRA: Extended Resource (#5136)
kubernetes/enhancements: keps/sig-scheduling/5004-dra-extended-resource/README.md

  `ResourceSlice` to provide resource capacity. Hence, it is of type (string, int64)
   on the consumption side, and list of devices with a common
  `ExtendedResourceName` on the capacity side.

With these additions in place, the DRA devices can be consumed by extended resource
requests, or by DRA resouce claims. The scheduler has everything it needs to support
the dynamic allocation of devices to requests made through extended resource and
resource claims. No static partition of resources between extended resources and
resource claims is needed. The kubelet and DRA driver has everything they need
to admit a pod and pass the allocated devices to the containers in the pod to run.

Note the following cluster setup configuration and constraint:

* One node in cluster can have an extended resource backed by DRA, and another node in the
cluster can have the same named extended resource backed by Device Plugin.

* One node in a cluster cannot have both extended resource backed by DRA, and same
  named extended resource backed by device plugin at the same time. This implies
  that either the resource is advertised through `ResourceSlice`,

>   or `Node`'s status.capacity.

Reviewer: In my opinion this is important enough to add to non-goals.

Reviewer: BTW, is this going to be enforced somehow?

Author: added it to the Non-Goals. It is not enforced, but my prototype implementation just picks the device plugin, and ignores the DRA when both are present on one node at the same time.

   resource backed by device plugin, in that, it has a string name, and a
   discrete countable quantity. Its capacity can be derived from DRA
   `ResourceSlice`, its consumption is specified through pod's extended
   resource request.
1. Introduce a field `ExtendedResourceName` to `DeviceClass` to allow cluster
   administrators to treat certain class of devices as an extended resource.
1. Introduce a special `ResourceClaim` object to keep track of device allocations. It is special only in the sense that it is created by the scheduler. No semantic changes are needed in the ResourceClaim API for it.
   for all extended resource requests backed by DRA for a pod. kube-scheduler
   uses DRA scheduling algorithm to fit pod's extended resource request to a
   node that advertises the extended resource in DRA `ResorceSlice` or extended
   resources backed by device plugin. When using DRA devices, it creates a
   special `ResourceClaim` for the pod with the allocation result recording
   which devices were picked. More details on this special `ResourceClaim`
   follow below.  When using extended resources advertised for a node by device

>    plugin, the existing resource tracking reserves them.

Author: 1. Yes, during filtering phase, all nodes will be checked to see if can be a candidate for the pod, a node can have either device plugin or DRA driver
2. Yes, the claim is only used by the dynamicresource plugin, which is soly for DRA. The other parts of the scheduler code is not concerned with this special claim.
3. No, the only change needed is the nodresource plugin, the change will allow it to pass the 'extended resource' the device plugin cannot satisfy to the dynamicresource plugin to try to satisfy.
4. Right, noderesource plugin is responsible for satisfying exteneded resource backed by device plugin, and dynamicresouce plugin is responsible for satisyfing extended resource backed by DRA
5. there is at most one speical resource claim created per pod (all extended resources backed by DRA use the same one claim)

>    plugin, the existing resource tracking reserves them.

Author: it is actually simpler to use one claim per pod, than one claim per type, in that it is less bookkeeping work in the schduler plugin knowing there is only one per pod. That said, one claim or one per type are implemantion detail, we can make the API PodExtendedResourceStatuses (instead of PodExtendedResourceStatus), and in Alpha, we use one claim per pod, if that turns out to be a wrong choice, we can switch to use one claim per type w/o changing the API.

>    plugin, the existing resource tracking reserves them.

Reviewer 2: > won't the scheduler plugin be significantly more efficient / simpler during the filtering pass keeping one claim per type?

No, I don't think so. The opposite might be true: there's more bookkeeping to be done when there are more claims.

> we can make the API PodExtendedResourceStatuses (instead of PodExtendedResourceStatus), and in Alpha, we use one claim per pod, if that turns out to be a wrong choice, we can switch to use one claim per type w/o changing the API.

If we allow multiple statuses, then all consumers of that API already now must support more than one. This might be the right choice (future-proof!), I just wanted to point out that it already makes the initial implementation more complex.

>    plugin, the existing resource tracking reserves them.

Reviewer 3: Let's defer the single-vs-plural decision to implementation time. If the performance and implementation complexity of the single claim turns out to be acceptable, and we have a clear view to the length limits of a single resource claim not being problematic long-term, we can keep the status singular as well.

With these additions in place, the DRA devices can be consumed by extended resource
requests, or by DRA resouce claims. The scheduler has everything it needs to support
the dynamic allocation of devices to requests made through extended resource and
resource claims. No static partition of resources between extended resources and
resource claims is needed. The kubelet and DRA driver has everything they need
to admit a pod and pass the allocated devices to the containers in the pod to run.

Note the following cluster setup configuration and constraint:

* One node in cluster can have an extended resource backed by DRA, and another node in the
cluster can have the same named extended resource backed by Device Plugin.

* One node in a cluster cannot have both extended resource backed by DRA, and same
  named extended resource backed by device plugin at the same time. This implies
  that either the resource is advertised through `ResourceSlice`,
  or `Node`'s status.capacity.

## Design Details

### Device Class API
The extended resource name to DRA device mapping can be specified at
`DeviceClassSpec`. The same extended resource name can be given to at most one 

> device class.

Author: during DeviceClass create & update, we can validate if there is another one with the same extended resource name.

extendedResourceName on DeviceClass is mutable. scheduler does the scheduling decision based on the snapshot of the scheduling cycle. during reattempt of scheduling a pod, scheduler will try to find if there is already a special claim created, and reuse it if it is still valid. Otherwise, it will be deleted, and a new one will be created.

> device class.

Reviewer 2: > during DeviceClass create & update, we can validate if there is another one with the same extended resource name.

Unfortunately we can't. Validation is limited to a single resource, the one which gets added or updated. We have to detect it during scheduling.

> device class.

Reviewer 3: > during DeviceClass create & update, we can validate if there is another one with the same extended resource name.

As @pohly mentioned, we cannot enforce that.

Describe what happens at scheduling if >1 exist that both reference the extended resource name.

> device class.

Author: https://github.com/kubernetes/enhancements/commit/807c8f336f8673a0816f64b8af8e3ab1cdb97f60

when >1 exists, one is randomly picked at scheduling time, i think this is better UX than failing it completely.

> device class.

Reviewer 3: Hmm... random is not ideal. Can we be deterministic, and also consider surfacing the situation somehow (event or condition or something)?

> device class.

Reviewer 3: compare what we do for other things:
* https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/storageclass.go#L58-L70
* https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/network/defaultingressclass/admission.go#L145-L153

> device class.

Author: that sounds better UX, reworded the KEP in https://github.com/kubernetes/enhancements/commit/97657b76a7ebf83a5b9fc9fa86628c15ae8e2b21

Cluster administrator is soly responsible for creating device classes, and the
mapping between the class of devices and the extended resource name.
`DeviceClass` is cluster scoped, application developers and operators cannot change it.

The mapping of DRA devices and extended resources is stored in k8s data store
(e.g. etcd). It is created after cluster creation, before deployment of the

> application that uses the devices.

Author: this ordering is required to use extended resource backed by DRA. However, migration can happen on a node by node basis.

0/ the cluster has all the nodes using device plugin (initial cluster state)
1/ cluster admin creates the device class with the 'extended resource name'
2/ the node is updated to switch from device plugin to DRA, one node a time
3/ new pod can be scheduled to the updated node with DRA

> application that uses the devices.

Reviewer 3: Since there are scenarios where applications can already exist which use extended resources before the device class is created, I think this statement is more confusing than helpful... maybe drop it?

> application that uses the devices.

Author: added commit hopefully to make it clearer https://github.com/kubernetes/enhancements/commit/2a5b1b604006a48d53cfadf38be6d6efb169011d

```go
// DeviceClassSpec is used in a DeviceClass to define what can be allocated
// and how to configure it.
type DeviceClassSpec struct {
	// ExtendedResourceName defines a mapping to the extended resource API.
	// All devices matched by the device class can be used to satisfy extended resource requests in pod's spec using this name.
	//
	// +optional
	ExtendedResourceName *string
}
```
#### Implicit Extended Resource Name

In addition to this optional extended resource name that is explicitly defined, every device class can be accessed
as an extended resource using the name `deviceclass.kubernetes.io/<device-class-name>`. This implicit extended
resource name allows the simpler API to be used for DRA resource when no special DRA features beyond those

> available via `DeviceClass` are needed.

Author: DeviceClasses is designed to be used only by cluster admin, app operator is not allowed to use it.

kube-scheduler created resource claim is special, it is not intened to be used by user. (we may add validation to disallow the speical claim from being used in a pod spec). it can only be used indirectly by the extended resource requests in Pod spec.

will check with pod-level resource folks if they intend to extend it beyond cpu/memory

> available via `DeviceClass` are needed.

Reviewer 3: > kube-scheduler created resource claim is special, it is not intened to be used by user. (we may add validation to disallow the speical claim from being used in a pod spec).

Be really careful about constructing "special" objects inside a namespace we can only detect via things that a user could also mess with (e.g. annotations). Assume a user will (intentionally or accidentally) create things indistinguishable from your special objects if they are able to.

> available via `DeviceClass` are needed.

Reviewer 3: > DeviceClasses is designed to be used only by cluster admin, app operator is not allowed to use it.

How does the cluster admin quota app operators' use of an extended resource referenced in a pod which *also* triggers creation of a resourceclaim using a device class? Today they can do so via regular pod resource quota (for extended resources) or device class quota, and there's no intersection between those two worlds ... this KEP will make usage show up in both places, it seems like we need a plan to not double-count in quota.

> available via `DeviceClass` are needed.

Reviewer 3: > will check with pod-level resource folks if they intend to extend it beyond cpu/memory

I guess if we're doing this implicit version of DRA to make a more usable version of explicit resource claims, and pod-level resources don't allow explicit resource claims, maybe we don't need to worry about that.

> available via `DeviceClass` are needed.

Author: For the "special" object, in addition to the special annotation (with resource.kubernetes.io/xxx), which user can mess with, but not supposed to. the object also has the owner reference's controller field set to true, currently, there is one controller, the resoure claim controller, and scheduler would be a 2nd one.

> available via `DeviceClass` are needed.

Author: For the quota, there would be two ways, one is through the extended resources quota, the other is through the DRA devices quota. both can be used, and work at the same time. the end result is whichever hits first takes the effect.

extended resource quota covers all extended resources usage, both from DRA and from device plugin.

DRA device quota covers all devices, for both the devices consumed by resourceclaim, and devices consumed from extended resources.

> available via `DeviceClass` are needed.

Reviewer 3: > For the "special" object, in addition to the special annotation (with resource.kubernetes.io/xxx), which user can mess with, but not supposed to. the object also has the owner reference's controller field set to true, currently, there is one controller, the resoure claim controller, and scheduler would be a 2nd one.

What are the consequences if a user sets that annotation or ownerReference? What behavior can they drive? Are there any validations or constraints they can escape?

> available via `DeviceClass` are needed.

Reviewer 3: > extended resource quota covers all extended resources usage, both from DRA and from device plugin.

Will this be a modification to the quota admission plugin and controller to teach it about device classes -> extended resource mappings?

> For the quota, there would be two ways, one is through the extended resources quota, the other is through the DRA devices quota. both can be used, and work at the same time. the end result is whichever hits first takes the effect.

That's my worry... I don't think it will be a "first hit" effect with DRA-backed extended resources. It seems like the pod spec will get counted AND the resulting resourceclaim will get counted. It also seems like we can't easily tell quota to ignore either, since you can use resourceclaims explicitly, and at pod admission time we don't know if DRA-backing will be used.

> available via `DeviceClass` are needed.

Author: For quota, both pod spec, and resulting resource claim are counted. if a cluster admin sets up just quota for extended resource, then the pod spec is used. if a cluster admin sets up just quota for resource claims, then only that claim quota is used. but if the cluster admin sets up both, then the first hit will limit the usage. (there could be the case that the schduler cannot create the special resourceclaim for the pod due to 'quota exceeded')

> available via `DeviceClass` are needed.

Reviewer 3: Scenario:
* Admin wants to limit to 10 GPUs which are addressable via extended resources OR DRA
  * create a deviceclass quota of 10
  * create an extended resource quota of 10
* Cluster has 50% of nodes with DRA / 50% with extended device plugins
* User:
  * Creates 10 pods with extended resources - uses *all* of their extended resource quota; only ~half result in ResourceClaims, which will use ~5 of the deviceclass quota
  * Can create 5 more pods using resourceclaims directly before reaching deviceclass quota limit
  * This effectively lets them use 15 GPUs

Solution:
* Update DRA `claimEvaluator#Usage` calculation of usage for ResourceClaims for DeviceClasses which map to an extended resource to *also* increment the extended resource?
  * This will double-count usage at the pod and at the resource claim; we *only* want to do this for direct use of ResourceClaims, not ones created for pods, which we can't reliably identify. Maybe we can follow the ResourceClaim → ownerReference → Pod → to determine if this is a "special" resource claim or not? If the pod has no references to the resourceClaims and no resourceClaimTemplates, we can assume the claim is "special"

> available via `DeviceClass` are needed.

Author: @pohly @johnbelamaric @liggitt  I think the resource quota controller can be used to reconcile the differences between the two quotas: extended resource quota, and resource claim quota. I added a section to the KEP, please take a look.

### Resource Claim API

A special resource claim object is created to keep track of device allocations for
extended resource. The resource claim object has the following properties:

  * It is namespace scoped, like other resource claim objects.
  * It is owned by a pod, like other resource claim objects.
  * It has `Spec` of device.requests, with each request name being an encoding
    of the container name and the extended resource backed by DRA name inside
    the container. It is needed to map allocated devices back to the container

>     which requested them.

Author: device request name is "container-%d-request-%d", the first %d is the container's index inside the pod, and the second %d is the extended resource's index within the container's requests. The request name is short.

the second statement is correct, ContainerExtendedResourceRequest provides the mapping. removed the first statement to avoid confusion.

>     which requested them.

Reviewer 3: > the first %d is the container's index inside the pod

if there are initContainers and containers, how is the indexing done?
- initContainer 0 == 0?
- initContainer 1 == 1?
- container 0 = 2?
- container 1 = 3?

> the second %d is the extended resource's index within the container's requests

index in sorted order among all the requested resources (whether extended or not) is simpler to describe... is that what you mean?

>     which requested them.

Author: i'm thinking to put all the container names into a list, then sort the list, then take the index from that sorted list.

For extended resources, you're right, it is to sort all the resources in a container's requests, then take the index for the extended resources.

>     which requested them.

Reviewer 3: I think it's much more intuitive to refer to the container by its actual index in the list, not its index in a list of the containers sorted by name. Container order in pods is immutable.

>   * Its `status.allocation.devices` and `status.allocation.reservedFor` are

Author: there are five create/update calls:
1/ create the special claim
2/ update claim status
3/ update pod status
4/ update claim's finalizer (before delete in PostFilter phase)
5/ update claim's finalizer (bfore delete in UnReserve phase)

there are two delete calls
11/ delete claim in PostFilter phase
12/ delete claim in UnReserve phase
13/ (GC delete claim when pod is deleted, this is not direct delete call in scheduler)

There are local retries within one scheduling cycle, and then there are retries across scheduliing cycle.

There is at most one such special claim per pod, scheduler needs to ensure this invariant, and either reuse it in subsquent scheduling cycle, or delete it and create a new one.

>   * Its `status.allocation.devices` and `status.allocation.reservedFor` are

Reviewer 3: The enumeration of the new API calls is helpful. Describe what happens before exiting the scheduling pass if each of those fail, and describe any cleanup done at the start of the next scheduling pass if there are remnants left that couldn't be cleaned up from the last scheduling pass.

    used.
  * It does not have annotation `resource.kubernetes.io/pod-claim-name:` as
    it is created for the extended resource request(s) in a pod spec, not for a
    claim in the pod spec.
  * It does have annotation `resource.kubernetes.io/extended-resource-claim: pod-name` as
    it is created, deleted, updated by the scheduler. It is used by scheduler
    to find the resource claim it has created, and ensure at most one such

>     claim per pod.

Reviewer 2: > Will it do that using live list lookups, or by checking an informer cache?

The plugin uses an assume cache, so it always knows about all claims that it created or updated, even if the underlying informer cache hasn't been updated yet through the watch.

>     claim per pod.

Author: we can use a label, instead of annotation. annotation is more common for use case like this, but label selector is more efficient.

Yes, the at most one is enforced by scheduler. assume cache is used to avoid the delay in informer sync

I'm not aware of the plan to adjust the 32 requests / 128 count limits in 1.34, I guess they can be adjusted upward if they turn out to be too restrictive. agreed that scheduler need to check these limits when trying to create the special claim for extended resource.

multiple resource claims are for claim (and resource) sharing across pod, and containers within a pod. The special claim for extened resource is not shared across pod, it is shared by all containers within a pod (kubelet has the logic to assign the devices allocated to the related containers)

>     claim per pod.

Reviewer 3: hmm... pod names must match IsDNS1123Subdomain (up to 253 characters with dots), label values must be <=63 characters and have a different regex, so you can't put arbitrary pod names into a label value.

Do we actually need the pod name in the annotation, or is the indication this is an extended resource claim sufficient in combination with the ownerReference to find the owning pod?

>     claim per pod.

Author: if we first List then go over the list one by one to check the owner reference, then we don't need to know the name. But if we want to use label selector, we need to know the name, that does not need to be the exact pod name, we can use a pod name hash (SHA-1) with some fixed prefix for that purpose.

>     claim per pod.

Reviewer 3: For what it's worth, label selector is also implemented as a linear search, it just happens deeper inside the informer lister or server-side. Needing to search through all claims for every pod this path handles could be really impactful in a namespace with with many thousands of pods / claims (whether done via an annotation search or a label selector) ... make sure we prove this isn't a problematic bottleneck

    preBind phase. The in-memory one in the assumed cache is created earlier
    during Reserve phase.
  * It is *deleted*
    * either together with the owning pod's deletion.
    * or by the scheduler dynamic resource plugin during unReserve phase.
    * or by the scheduler dynamic resource plugin during postFilter phase.
  * It is *discovered* by the kubelet via `pod.Status.ExtendedResourceClaimStatus`
  * It is *read* by the kubelet DRA device driver to prepare the devices listed
    therein when preparing to run the pod.

```go
type DeviceRequest struct {
	// Name can be used to reference this request in a pod.spec.containers[].resources.claims
	// entry and in a constraint of the claim.
	//
	// Must be a DNS label.
	//
	// +required
	Name string
}
```

To enable the kubelet to map devices back to the containers which requested them,
the kube-scheduler creates one `DeviceRequest` per extended resource backed by DRA
per container in the pod. The name of the `DeviceRequest` has the form
"container-%d-request-%d", where the first %d is the index of the container in the pod.
The second %d is the index of the extended resource inside the container

> resource requests. For example, if the first container in the pod has an

Author: all containers in the pod are counted for the indexing, sorting is based on container names

all resources in a container are counted for the indexing, sorting is based on resource names

scheduler (dynamic resource plugin) is the only component that does this indexing and sorting. The mapping is recorded in pod.status.ExtendedResourceClaimStatus, so that kubelet just need to do the loop up by container name, and resource name to get the request name.

> resource requests. For example, if the first container in the pod has an

Reviewer 3: > all containers in the pod are counted for the indexing, sorting is based on container names

so even though containers are in a list, we sort them by name? that's confusing

> all resources in a container are counted for the indexing, sorting is based on resource names

+1

> resource requests. For example, if the first container in the pod has an

Author: >> so even though containers are in a list, we sort them by name? that's confusing

do we allow reordering of the containers in the list? if this ordering is not guaranteed for now, and future, then sorting is safer. the index used  in the request name is not really meaningful to user, user should check the pod.status.extendedResourceClaimStatus for the mapping from container name, extended resource name to the request name

> resource requests. For example, if the first container in the pod has an

Reviewer 3: > do we allow reordering of the containers in the list?

no

### Pod API

A new field `extendedResourceClaimStatus` is added to Pod's status to track
the special `ResourceClaim` object created for the extended resource requests
in the pod. This is needed for kubelet to pass the devices allocated by driver
to the containers in the pod.

```go
// PodExtendedResourceClaimStatus is stored in the PodStatus for each extended
// resource requests backed by DRA. It stores the generated name for 
// the corresponding special ResourceClaim created by scheduler.
type PodExtendedResourceClaimStatus struct {
        // ResourceClaimName is the name of the ResourceClaim that was
        // generated for the Pod in the namespace of the Pod.
        ResourceClaimName string

        // RequestMapping identifies the mapping of <container, extended resource backed by DRA> to  device request.
        // +patchMergeKey=requestName
        // +patchStrategy=merge,retainKeys
        // +listType=atomic
        // +listMapKey=requestName
        // +featureGate=DynamicResourceAllocation
        RequestMapping []ContainerExtendedResourceRequest
}

type ContainerExtendedResourceRequest struct {
        // ContainerName is the unique container name within the pod.

>         ContainerName string

Author: https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#detailed-behavior

The name of each app and init container in a Pod must be unique; a validation error is thrown for any container sharing a name with another.

        // ExtendedResourceName is the extended resource name backed by DRA inside
        // the container's requests.
        ExtendedResourceName string
        // RequestName is the device request name in the special resource claim
        // created for extended resource requests backed by DRA.
        RequestName string
}

type PodStatus struct {
    ...

    // Status of extended resource claim backed by DRA.
    // +featureGate=DynamicResourceAllocation
    // +optional

>     ExtendedResourceClaimStatus *PodExtendedResourceClaimStatus

Author: a resourceclaim is a list of DeviceClaims, and each DeviceClaim is to request for a type of device. It is supported to have requests for multiple types of devices in one claim.

>     ExtendedResourceClaimStatus *PodExtendedResourceClaimStatus

Reviewer 2: I don't expect the deepcopy to have a performance impact. I don't like it conceptually because it is less obvious than creating a new special claim from scratch for each node which needs it.

extended resource request is checked against a NodeInfo's ScalarResources if the
node uses device plugin, and checked against a NodeInfo's DynamicResources, if
the request is for extended resources backed by DRA, then 'noderesources' plugin
would pass and leave it to 'dynamicresource' plugin to check if it can be satisfied.

The existing 'dynamicresources' plugin needs to be modified to account for the
extended resource backed by DRA requests.

#### EventsToRegister
This registers all cluster events that might make an unschedulable pod schedulable,
like finishing the allocation of a claim, or resource slice updates.

The existing dynamicresurce plugin has registered all the events needed or
extended resource backed by DRA, hence no change is needed.

#### PreFilter
It checks if the pod has any container requests for extended resources backed by DRA.
If not, and no claims in the pod, then the plugin can return early, as there
is nothing to do.

If the pod still needs to be considered by the plugin, then it checks if the
special resource claim for extended resources backed by DRA has been created
before by scheduler, by checking resource claim name having pod name in the

> annotation `resource.kubernetes.io/extened-resource-claim: pod-name`.

Author: the plugin uses assume cache, instead of relying on the informer cache which could be stale.

a resoureclaim is created with a nodeselector if it is allocated, in that case, the scheduler will try to reuse it, and if it does not work, then during PostFilter phase, it will be deleted, and next scheduling cycle will create a new one.

> If found, scheduler would reuse it. If not found, scheduler would create a

Author: "create" means in-memory, the API call to delete happens during the later 'PreBind' phase. scheduler knows whether the claim is in-memory claim (a temporary one), or a real claim (persisted already on API server)

special resource claim that has empty spec. The exact spec needs to be decided
during Filter phase, as some node may have device plugin provide the capacity
for the extended resource, some other node may have DRA provide the capacity.
The requests in the special resource claim need to vary for each node.

#### Filter
If a pod has an extended resource backed by DRA, and the node does *not* have
device plugin to provide the capacity for the resource, then the
dynamicresource plugin needs to try allocate the resource by filling in the

> special claim's `Spec.Devices.Requests` field.

Author: Yes, it is a separate copy per node.

I think it's simpler to have at most one special claim per pod. if scheduler has created the claim, but failed somehow, then at the preFilter phase of the next scheduling cycle, it will validate the claim to see if it can be reused, if not, (say device class and extended resource mapping has changed), then the claim needs to be deleted, and a new one needs to be created.

> special claim's `Spec.Devices.Requests` field.

Reviewer 3: hmm... maybe there's a way to structure the dynamic scheduler plugin to simplify construction of a claim-per-node rather than making a complete one and deepcopying it for every node

> special claim's `Spec.Devices.Requests` field.

Reviewer 3: we can discuss this more in the implementation, but there's more reasons than performance to avoid the "construct once, then deepcopy and mutate" approach ... it ends up being really difficult to reason about

> special claim's `Spec.Devices.Requests` field.

Author: sounds good, let's discuss more in-depths in the implementation PR

Note the following cluster setup configuration and constraint:

* One node in cluster can have an extended resource backed by DRA, and another node in the
cluster can have the same named extended resource backed by Device Plugin.

* One node in a cluster cannot have both extended resource backed by DRA, and same
  named extended resource backed by device plugin at the same time. This implies
  that either the resource is advertised through `ResourceSlice`,
  or `Node`'s status.capacity.

## Design Details

### Device Class API
The extended resource name to DRA device mapping can be specified at
`DeviceClassSpec`. The same extended resource name can be given to at most one 
device class.

Cluster administrator is soly responsible for creating device classes, and the
mapping between the class of devices and the extended resource name.
`DeviceClass` is cluster scoped, application developers and operators cannot change it.

The mapping of DRA devices and extended resources is stored in k8s data store
(e.g. etcd). It is created after cluster creation, before deployment of the
application that uses the devices.

```go

> // DeviceClassSpec is used in a [DeviceClass] to define what can be allocated

Author: removed markup

  * It is owned by a pod, like other resource claim objects.
  * It has `Spec` of device.requests, with each request name being an encoding
    of the container name and the extended resource backed by DRA name inside
    the container. It is needed to map allocated devices back to the container
    which requested them.
  * Its `status.allocation.devices` and `status.allocation.reservedFor` are
    used.
  * It does not have annotation `resource.kubernetes.io/pod-claim-name:` as
    it is created for the extended resource request(s) in a pod spec, not for a
    claim in the pod spec.
  * It does have annotation `resource.kubernetes.io/extended-resource-claim: pod-name` as
    it is created, deleted, updated by the scheduler. It is used by scheduler
    to find the resource claim it has created, and ensure at most one such
    claim per pod.
  * At most one such claim object is created per pod. For example, if a pod
    requests for foo1.domain/bar and foo2.domain/bar, the allocation of devices
    for each are recorded in DeviceResourceRequestAllocationResult, and just 
    one claim object with allocation `Results` that lists all allocated devices

>     is created for the pod.

Author: added the failure mode to the troubleshooting section

The special resource claim object lifecycle is managed by the scheduler and
garbage collector.

  * It is *created* in a namespace when there is a pod with extended resource
  request, and the extended resource is advertised by `ResourceSlice` and
  scheduler has fit the pod to a node with the `ResourceSlice`.
  * It is *created* by the scheduler dynamic resource plugin during
    preBind phase. The in-memory one in the assumed cache is created earlier
    during Reserve phase.
  * It is *deleted*
    * either together with the owning pod's deletion.
    * or by the scheduler dynamic resource plugin during unReserve phase.
    * or by the scheduler dynamic resource plugin during postFilter phase.
  * It is *discovered* by the kubelet via `pod.Status.ExtendedResourceClaimStatus`
  * It is *read* by the kubelet DRA device driver to prepare the devices listed
    therein when preparing to run the pod.

```go
type DeviceRequest struct {
	// Name can be used to reference this request in a pod.spec.containers[].resources.claims
	// entry and in a constraint of the claim.
	//
	// Must be a DNS label.
	//
	// +required

> 	Name string `json:"name" protobuf:"bytes,1,name=name"`

Reviewer 2: ```suggestion
	Name string
```

Same elsewhere. Please also replace tab with four spaces, for consistency and easier editing.

Author: done

}
```

To enable the kubelet to map devices back to the containers which requested them,
the kube-scheduler creates one `DeviceRequest` per extended resource backed by DRA
per container in the pod. The name of the `DeviceRequest` has the form
"container-%d-request-%d", where the first %d is the index of the container in the pod.
The second %d is the index of the extended resource inside the container
resource requests. For example, if the first container in the pod has an
extended resource backed by DRA which is the 3rd such request in the container,
then the name of the `DeviceRequest` is "container-0-request-2".

Documenting this naming is merely informational, it is not part of the API.
The kubelet must not rely on it. Instead, the
`ContainerExtendedResourceRequest` field below specifies the mapping.

### Pod API

> A new field `extendedResourceClaimStatus` is added to Pod's status to track

Reviewer 2: I like how well "Extended Resource" and "ResourceClaim" combine into "ExtendedResourceClaimStatus" here :smile: 

"ExtendedResourceDeviceClaim"... not so much.

@yliaog: this refers to some old discussions around renaming ResourceClaim to DeviceClaim.

the special `RresouceClaim` object created for the extended resource requests
in the pod. This is needed for kubelet to pass the devices allocated by driver
to the containers in the pod.

```go
// PodExtendedResourceClaimStatus is stored in the PodStatus for each extended
// resource requests backed by DRA. It stores the generated name for 
// the corresponding special ResourceClaim created by scheduler.
type PodExtendedResourceClaimStatus struct {
        // Names identifies the mapping of <container, extended resource backed by DRA> to  device request.
        // +patchMergeKey=requestName
        // +patchStrategy=merge,retainKeys
        // +listType=atomic
        // +listMapKey=requestName

>         // +featureGate=DynamicResourceAllocation

Reviewer 2: Wrong feature gate name, also below.

I'm not sure about the "Names" field name. It's too vague and doesn't describe the semantic.

```yaml
   names:
   - containerName: my-container
     extendedResourceName: dra.example.com/gpu
     requestName: container-0-request-2
```

How about `RequestMapping`?

        Names []ContainerExtendedResourceRequest `json:"names" patchStrategy:"merge,retainKeys" patchMergeKey:"requestName" protobuf:"bytes,1,rep,name=names"`
           
        // ResourceClaimName is the name of the ResourceClaim that was
        // generated for the Pod in the namespace of the Pod.

>         ResourceClaimName string `json:"resourceClaimName" protobuf:"bytes,2,name=resourceClaimName"`

Reviewer 2: Nit: I think this should come first. It feels like the more important field.

Ordering in kubectl unfortunately will have `resourceClaimName` after `requestMapping`.

Author: reordered, put ResourceClaimName before requestMapping

}

type ContainerExtendedResourceRequest struct {
        // ContainerName is the unique container name within the pod.
        ContainerName string `json:"containerName" protobuf:"bytes,1,name=containerName"`
        // ExtendedResourceName is the extended resource name backed by DRA inside
        // the container's requests.
        ExtendedResourceName string `json:"extendedResourceName" protobuf:"bytes,2,name=extendedResourceName"`
        // RequestName is the device request name in the special resource claim
        // created for extended resource requests backed by DRA.
        RequestName string `json:"requestName" protobuf:"bytes,3,name=requestName"`
}

> // PodStatus represents information about the status of a pod. Status may trail the actual
> // state of a system, especially if the node that hosts the pod cannot contact the control
> // plane.
> type PodStatus struct {
>         // Status of resource claims.
>         // +patchMergeKey=name
>         // +patchStrategy=merge,retainKeys
>         // +listType=map 
>         // +listMapKey=name
>         // +featureGate=DynamicResourceAllocation
>         // +optional
>         ResourceClaimStatuses []PodResourceClaimStatus `json:"resourceClaimStatuses,omitempty" patchStrategy:"merge,retainKeys" patchMergeKey:"name" protobuf:"bytes,15,rep,name=resourceClaimStatuses"`
>         // Status of extended resource claim backed by DRA.
>         // +featureGate=DynamicResourceAllocation
>         // +optional
>         ExtendedResourceClaimStatus *PodExtendedResourceClaimStatus `json:"extendedResourceClaimStatus,omitempty" protobuf:"bytes,17,opt,name=extendedResourceClaimStatus"`
> }

Reviewer 2: ```suggestion
type PodStatus struct {
    ...

    // Status of extended resource claim backed by DRA.
    // +featureGate=DynamicResourceAllocation
    // +optional
    ExtendedResourceClaimStatus *PodExtendedResourceClaimStatus
}
```

```

For example, if a pod has requested for foo.domain/bar, and it is
scheduled to run on a node where foo.domain/bar was mapped to devices in a DeviceClass,
then the pod's status is like below:

```yaml
status:
  extendedResourceClaimStatus:

>     names:

Reviewer 2: requestMapping?

Author: renamed.

    - containerName: container-name
      extendedResourceName: foo.domain/bar
      requestName: container-0-request-2

>     resourceClaimName: ccc-gpu-57999b9c4c-vpq68-gpu-8s27z

Reviewer 2: Put this first?

Author: done

   for all extended resource requests backed by DRA for a pod. kube-scheduler
   uses DRA scheduling algorithm to fit pod's extended resource request to a
   node that advertises the extended resource in DRA `ResorceSlice` or extended
   resources backed by device plugin. When using DRA devices, it creates a
   special `ResourceClaim` for the pod with the allocation result recording
   which devices were picked. More details on this special `ResourceClaim`
   follow below.  When using extended resources advertised for a node by device
   plugin, the existing resource tracking reserves them.
1. Introduce a field `ExtendedResourceClaimStatus` to pod's `Status`, such that:
    - the kubelet can find the special `ResourceClaim` while looking for claims to prepare
    - the kubelet can pass the devices to containers in the pod with the extended
   resource requests, based on the container/extended resource to device
   request mapping in the `ExtendedResourceClaimStatus`.

Some quick clarifications around the basic concepts: extended resource backed by
device plugin, extended resource backed by DRA, and dynamic resource.

* extended resource backed by device plugin uses pod's

>   spec.containers[].resources.requests to request for resources, it consumes the capacity

Reviewer 3: @tallclair, does IPPR allow mutation of extended resources in `spec.containers[].resources.requests`?

Since ResourceClaims are immutable, it doesn't seem like there's a reasonable way to honor resize requests on DRA-backed extended resources.

What would happen with those resize requests on the kubelet with device-plugin backed things today?

Can we make sure the DRA-backed ones respond in a way that is not *worse* than what device-plugin backed things do?

Reviewer 3: Nevermind, I just swept the IPPR validation and it limits resize to CPU/memory 👍 

@tallclair, can you make an explicit note in the IPPR design that we should consider this interaction if there was ever a desire to let extended resources be mutable.

>   spec.containers[].resources.requests to request for resources, it consumes the capacity

Reviewer 3: anywhere this KEP references `spec.containers[]...` can we make it clear that initContainers are included as well? (but not ephemeralContainers, I think, since those can't have resources and are added post-scheduling)

Author: clarified in the commit: https://github.com/kubernetes/enhancements/pull/5136/commits/06a7dac85a1453cfd3c2187650d5ab79af00d7b4

at scheduling time, if two are created at the same time, the name
lexicographically sorted first is picked.

Cluster administrator is soly responsible for creating device classes, and the
mapping between the class of devices and the extended resource name.
`DeviceClass` is cluster scoped, application developers and operators cannot change it.

The mapping of DRA devices and extended resources is stored in k8s data store
(e.g. etcd). An application using the extended resources can only request the
devices from DRA after the device class with the mapping is created. Before
that, the application can request the devices from device plugin only.

```go
// DeviceClassSpec is used in a DeviceClass to define what can be allocated
// and how to configure it.
type DeviceClassSpec struct {
	// ExtendedResourceName defines a mapping to the extended resource API.
	// All devices matched by the device class can be used to satisfy extended resource requests in pod's spec using this name.
	//
	// +optional
	ExtendedResourceName *string
}
```
#### Implicit Extended Resource Name

In addition to this optional extended resource name that is explicitly defined, every device class can be accessed

> as an extended resource using the name `deviceclass.kubernetes.io/<device-class-name>`. This implicit extended

Author: https://github.com/kubernetes/enhancements/pull/5136/commits/3e986dcaf7b04e7cc83146ca328d8a8d86a49f59

node uses device plugin, and checked against a NodeInfo's DynamicResources, if
the request is for extended resources backed by DRA, then 'noderesources' plugin
would pass and leave it to 'dynamicresource' plugin to check if it can be satisfied.

The existing 'dynamicresources' plugin needs to be modified to account for the
extended resource backed by DRA requests.

#### EventsToRegister
This registers all cluster events that might make an unschedulable pod schedulable,
like finishing the allocation of a claim, or resource slice updates.

The existing dynamicresource plugin has registered all the events needed or
extended resource backed by DRA, hence no change is needed.

#### PreFilter
It checks if the pod has any container requests for extended resources backed by DRA.
If not, and no claims in the pod, then the plugin can return early, as there
is nothing to do.

If the pod still needs to be considered by the plugin, then it checks if the
special resource claim for extended resources backed by DRA has been created
before by scheduler, by checking resource claim name having pod name in the
annotation `resource.kubernetes.io/extended-resource-claim: pod-name`.

> If found, scheduler would reuse it. If not found, scheduler would create a

Reviewer 4: Please clarify this "create" means just in-memory thingy on the KEP.
I was about to make the same question as @/liggitt here before I saw:
https://github.com/kubernetes/enhancements/pull/5136/files#r2091951508

Reviewer 4: And, also how is this in-memory ResourceClaim implemented? Through the scheduler cache?

    - Detection: How can it be detected via metrics? Stated another way:
      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->
  - [Pod pending due to extended resource backed by DRA requests no less than 128 devices]
    - Detection: inspect pod status 'Pending'
    - Mitigations: reduce the number of devices requested in one extended resource backed by DRA requests
    - Diagnostics: scheduler logs at level 5 show the reason for the scheduling failure.
    - Testing: Will be considered for beta.

###### What steps should be taken if SLOs are not being met to determine the problem?

Will be considered for beta.

## Implementation History

- Kubernetes 1.34: KEP accepted.

## Drawbacks

It adds complexity to the scheduler.

## Alternatives

Many different approaches were considered.

> Specifically, the following two alternative proposals were considered:

Reviewer 4: I don't fully understand a strong enough reason why this has to be implemented in kube-scheduler, i.e., during the scheduling flow. 

Have we considered this: gate the pods with the extended resource request at kube-apiserver, create a separate controller making necessary API calls, and un-gate pods. 
(Basically same as what I'm suggesting at the implicit tolerations discussion: https://github.com/kubernetes/enhancements/issues/5282#issuecomment-2901506804.)

**Option 1:** webhook rewrite extended resource requests in pod spec

This approach requires cluster administrator deploy a mutation webhook to the
cluster, and configure the webhook with rewrite rules that can rewrite the
extended resource requests and node selectors. This approach is not taken due to

> the webhook's extra configuration, and maintenance overhead.

Reviewer 4: Can't it be just done at kube-apiserver side so that there'll be no install/config on the users' side? Though I'm not sure there's been any cases that creates another resource when receiving requests at kube-apiserver. (like this case, we need to create ResourceClaim at Pod creation)

Author: the exact resourceclaim's requests cannot be determined until at scheduling Filter phase inside dynamic resources plugin. kube-apiserver does not have that necessary information to create the requests for the resourceclaim. dynamicresource plugin is doing the DRA resource allocations, it has all the information.

at most one device class). The `Name` of the request is determined by the
container name and the extended resource name.

The allocator needs to be modified to allow for the special resource claim for
extended resource backed by DRA, which could vary by node. The `Allocate`
method takes the claim as a parameter, in adddition to node parameter. The
algorithm uses the passed in special claim whenever it processes the last claim
in the claims slice, which is an instantiation of the special claim template
created during the preFilter phase.

If there is an allocation for a node, the allocation, and the claim are
recorded in cyclestate.

#### PostFilter
If the special resource claim is not available, i.e., the claim cannot be bound
to the node, then scheduler would deallocate it, and delete it during
PostFilter phase.

#### Reserve
Reserve the in-memory `ResourceClaim` and its allocation results in the assume
cache, a map of in-flight claims.

#### Unreserve
The plugin  removes the allocation for the special `ResourceClaim` for extended
resource backed by DRA. Because it cannot be scheduled after all. It does not need

> to remove the Pod from the claim.status.reservedFor field, as the special claim

Reviewer 5: I think it would be better to clear reservedFor if there are no other reasons behind it. This detail could be also left for implementation, as I don't see any reason to specify it in the KEP.

Author: there is indeed one "optional" resource claim that is created in-memory in preFilter, but that claim is just a template, it has to be instantiated at the Filter stage as the extended resource may be satisfied by device plugin in one node, but satisfied by DRA in another node, hence the claim has to vary per node. If a node has device plugin for that extended resource, then the claim does not need to request to allocate it again through DRA.

For Unreserve, I revised it such that Unreserve would delete the whole special claim (not just the allocation), as at the next scheduling cycle, the special claim may need to be different, it is better to clean it up, and restart from scratch for the next scheduling cycle.

    it is created, deleted, updated by the scheduler. It is used by scheduler
    to find the resource claim it has created, and ensure at most one such
    claim per pod.
  * At most one such claim object is created per pod. For example, if a pod
    requests for foo1.domain/bar and foo2.domain/bar, the allocation of devices
    for each are recorded in DeviceResourceRequestAllocationResult, and just 
    one claim object with allocation `Results` that lists all allocated devices
    is created for the pod.

The special resource claim object lifecycle is managed by the scheduler and
garbage collector.

  * It is *created* in a namespace when there is a pod with extended resource
  request, and the extended resource is advertised by `ResourceSlice` and
  scheduler has fit the pod to a node with the `ResourceSlice`.
  * It is *created* by the scheduler dynamic resource plugin during
    preBind phase. The in-memory one in the assumed cache is created earlier
    during Reserve phase.
  * It is *deleted*
    * either together with the owning pod's deletion.
    * or by the scheduler dynamic resource plugin during unReserve phase.

>     * or by the scheduler dynamic resource plugin during postFilter phase.

Reviewer 6: This sounds complicated. Can the object be always around, and simply ignored if scheduled on a Device Plugin node for now? As we will eliminate nodes with the Device Plugin over time, keeping this complicated logic will not be required, but we will have to do it for back compat.

Author: we discussed this idea with the sig scheduling folks, that would mean some other controller needs to create the claim. we agreed to further explore that before graduating to Beta, together with the 'nomindation' concept the sig scheduling is working on.

Reviewer 5: We can keep creating RC by scheduler for alpha, but internally we could consider having such object even during implementation, as it could bring a simplification already.

"container-%d-request-%d", where the first %d is the index of the container in the pod.
The second %d is the index of the extended resource inside the container
resource requests. For example, if the first container in the pod has an
extended resource backed by DRA which is the 3rd such request in the container,
then the name of the `DeviceRequest` is "container-0-request-2".

Documenting this naming is merely informational, it is not part of the API.
The kubelet must not rely on it. Instead, the
`ContainerExtendedResourceRequest` field below specifies the mapping.

### Pod API

A new field `extendedResourceClaimStatus` is added to Pod's status to track
the special `ResourceClaim` object created for the extended resource requests
in the pod. This is needed for kubelet to pass the devices allocated by driver
to the containers in the pod. containers can be initContainers, regular containers,
but cannot be ephemeral containers. 

```go
// PodExtendedResourceClaimStatus is stored in the PodStatus for each extended
// resource requests backed by DRA. It stores the generated name for 
// the corresponding special ResourceClaim created by scheduler.

> type PodExtendedResourceClaimStatus struct {

Reviewer 6: Can we re-use this field: https://github.com/kubernetes/enhancements/blob/b869b5ce260577c14173a0649ff69e18c5ad3ffd/keps/sig-node/4680-add-resource-health-to-pod-status/README.md?plain=1#L226

  for the short term. Its ease of use is one big advantage to keep it remaining
  useful for the long term.

* Device plugin API must not change. The existing device plugin drivers must
  continue working without change.

* DRA driver API must not change. Core Kubernetes (kube-scheduler, kubelet) is
  preferred over DRA driver for any change needed to support the feature.

* Keep advertising only extended resources backed by device plugin in `node.status.Capacity`
  for Alpha. It will be revisited for Beta, based on Alpha feedback.

### Non-Goals

* Minimize kubelet or kube-scheduler changes. The feature requires necessary
  changes in both scheduling and actuation.

* One node has both extended resource backed by DRA, and the same named
  extended resource backed by device plugin at the same time.

## Proposal

The basic idea is the following:

1. Introduce an `extended resource backed by DRA` concept. It is like the current extended
   resource backed by device plugin, in that, it has a string name, and a
   discrete countable quantity. Its capacity can be derived from DRA
   `ResourceSlice`, its consumption is specified through pod's extended
   resource request.

> 1. Introduce a field `ExtendedResourceName` to `DeviceClass` to allow cluster

Reviewer 7: yeah, but cluster admin will install manifest from provider foo that adds extended resource mapping for device plugin bar, are we ok with asking all cluster operators to review all the drivers provided manifests?

> 1. Introduce a field `ExtendedResourceName` to `DeviceClass` to allow cluster

Reviewer 7: answered in https://github.com/kubernetes/enhancements/pull/5136#discussion_r2136587610

With these additions in place, the DRA devices can be consumed by extended resource
requests, or by DRA resouce claims. The scheduler has everything it needs to support
the dynamic allocation of devices to requests made through extended resource and
resource claims. No static partition of resources between extended resources and
resource claims is needed. The kubelet and DRA driver has everything they need
to admit a pod and pass the allocated devices to the containers in the pod to run.

Note the following cluster setup configuration and constraint:

* One node in a cluster can have an extended resource backed by DRA, and another node in the
same cluster can have the same named extended resource backed by Device Plugin.

* One node in a cluster cannot have both extended resource backed by DRA, and same
  named extended resource backed by device plugin at the same time. This implies
  that either the resource is advertised through `ResourceSlice`,
  or `Node`'s status.capacity.

## Design Details

### Device Class API
The extended resource name to DRA device mapping can be specified at
`DeviceClassSpec`. The same extended resource name should be given to at most one 

> device class. If there are more than one device classes, the one created later is picked 

Author: no, it is not immutable. what heuristic are you referring to?

> device class. If there are more than one device classes, the one created later is picked 

Reviewer 7: wonder if making this field immutable once set makes this simpler to reason about, or if is a "do not worry" case

> device class. If there are more than one device classes, the one created later is picked 

Author: i felt immutable is a bit restrictive. cluster admin may make a mistake and want to correct it, or want to allow their developers to use a different extended resource name for the device class.

cluster admin has to understand if they change the extended resource name mapping in the device class, then the old extended resource name cannot be used for requesting devices after a certain delay, and new extended resource name can be used for requesting devices after a certain delay, there is a short period of time that either the old name is used by the scheduler, or the new name. The allocated claim has the device requests, whose name is the encoding of the container and extended resource name in the pod, so exactly whether the old name, or the new name is used can be determined (but up to the scheduler to decide during the short period of transition from old to new name).

> device class. If there are more than one device classes, the one created later is picked 

Reviewer 2: The problem with immutable fields in a DeviceClass doesn't actually simplify anything because it can always just get deleted and recreated differently to get the same effect, so all consumers have to be prepared for the content of a DeviceClass to change.

> device class. If there are more than one device classes, the one created later is picked 

Reviewer 3: Letting it be mutable allows transitioning which deviceclass is used to back an extended resource more easily.

It's a good point to make sure that we don't rely on the deviceclass extended resource field lookup beyond the point of admission / scheduling, since it is mutable. After the point of scheduling, the resourceclaim / pod status is what tells the kubelet how to satisfy the extended resource for the pod, and is unaffected by mutations to the deviceclass field.

at scheduling time, if two are created at the same time, the name
lexicographically sorted first is picked.

Cluster administrator is soly responsible for creating device classes, and the
mapping between the class of devices and the extended resource name.
`DeviceClass` is cluster scoped, application developers and operators cannot change it.

The mapping of DRA devices and extended resources is stored in k8s data store
(e.g. etcd). An application using the extended resources can only request the
devices from DRA after the device class with the mapping is created. Before
that, the application can request the devices from device plugin only.

```go
// DeviceClassSpec is used in a DeviceClass to define what can be allocated
// and how to configure it.
type DeviceClassSpec struct {
	// ExtendedResourceName defines a mapping to the extended resource API.
	// All devices matched by the device class can be used to satisfy extended resource requests in pod's spec using this name.
	//
	// +optional

> 	ExtendedResourceName *string

Reviewer 7: related to https://github.com/kubernetes/enhancements/pull/5136/files#r2136536134 , should we ensure the `ExtendedResourceName` matches the current device driver name to avoid problems fo dra drivers taken on other vendors resources?

Author: i think we can enforce the tld matches, .e.g requires both the extended resource (e.g. xyz.example.com/gpu) and driver (abc.example.com) have example.com

Reviewer 7: that sounds a good idea that solves this problem

Reviewer 7: Patrick, to be clear, I do not have a strong opinion and people here may have more use cases I'm not aware, but I feel that providing much flexibility to a mechanism that should be only solving the migration from device plugins to dra drivers, can make the scope creeping and we end in a situation like python2 -> python2 or ipv4->ipv6 that nobody nevers moves from the legacy mechanism

Reviewer 7: resolving :)

enhanced to have it keep track of the devices from device plugin too.

As a device can be requested by resource claim, or by extended resource, the
cluster admin MUST create two quotas with the same limit on one class of devices
to effectively quota the usage of that device class.

For example, a cluster admin plans to allow 10 example.com/gpu devices in a
given namespace, they MUST create the following:

```yaml
apiVersion: v1
kind: ResourceQuota
metadata:
  name: gpu
spec:
  hard:
    requests.example.com/gpu: 10
    gpu.example.com.deviceclass.resource.k8s.io/devices: 10
```

Provided that the device class gpu.example.com is mapped to the extended
resource example.com/gpu.
```yaml
apiVersion: resource.k8s.io/v1beta1
kind: DeviceClass
metadata:
  name: gpu.example.com
spec:
  extendedResourceName: example.com/gpu
```

Resource Quota controller reconciles away the differences if any between the
usage of the two quota, and ensures their usage are always kept the same. For
that, the controller needs to have the permission to list the device classes
in the cluster to establish the mapping between device class and extended

> resource.

Reviewer 7: can this be problematic on clusters with asymetric deployments, where are nodes only with the legacy driver?

Author: in the case that all nodes are having device plugin, then all devices need to be requested through extended resource, like today, no resource claim can be satisfied (since there are no resource slices). in this case, only the extended resource quota takes effect, user can create claims, but the claim cannot be allocated with any device, so the claim is not consuming any quota.

user can still set up both extended resource quota, and claim quota, and both are showing the same usage, thoug in reality, there is no device allocated via claim.

will be instantiated at Filter phase, persisted at PreBind phase.

During Filter phase, if the special resource claim allocation's
node selector does not match, then return
[framework.Unschedulable](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/interface.go#L208),
then the resourceclaim may be picked for deletion during PostFilter phase.

### Actuation for Extended Resource backed by DRA
When a pod with extended resources requests is picked up by the kubelet on the
node it is scheduled to run, the following are particularly important:

1. Kubelet tries to admit the pod, the pod's extended resources requests
   should not be checked against the `Node`'s allocatable, as the resources are
   in `ResourceSlice`, not in `Node`. In reality, the current predicate.go has
   already removed the missing extended resources from node info for
   cluster-level resources, hence there is no extra logic needed to admit the
   extended resources backed by DRA.

1. Kubelet (DRA manager) passes the special `ResoureClaim` to DRA driver to
   prepare the devices, in the same way as that for normal `ResourceClaim`.

> 1. Kubelet passes the device IDs through CDI to the containers with the

Author: currently DRA already requires CDI, so it is not new requirement from this feature.

> 1. Kubelet passes the device IDs through CDI to the containers with the

Reviewer 7: discussed below, no need to duplicate discussion

- All issues and gaps identified as feedback during beta are resolved

### Upgrade / Downgrade Strategy

The usual Kubernetes upgrade and downgrade strategy applies for in-tree
components. Vendors must take care that upgrades and downgrades work with the
drivers that they provide to customers.

### Version Skew Strategy

All of the API extensions proposed in this KEP is the optional
`ExtendedResourceName` in `DeviceClass`, and `ExtendedResourceClaimStatus` in
`Pod`. There is no risk for version skew downgrades
because these `DeviceClass` and `Pod` will never have existed in
older clusters.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: DRAExtendedResource
  - Components depending on the feature gate:
    - kube-apiserver
    - kube-scheduler
    - kubelet

###### Does enabling the feature change any default behavior?

No

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Applications that were already deployed and are running will continue to

> work. They will continue to work when restarted because the CDI devices that

Author: Yes, it requires CDI, which is not new, as DRA already has such requirement.

> work. They will continue to work when restarted because the CDI devices that

Reviewer 7: DRA does not require CDI, it allows drivers to use it, but is not required

> work. They will continue to work when restarted because the CDI devices that

Author: interesting, from the code, it seems CDI is required.
https://github.com/kubernetes/kubernetes/blob/387f474920c40aa6d8132e41579a9066ced336ad/pkg/kubelet/cm/dra/manager.go#L363

> work. They will continue to work when restarted because the CDI devices that

Reviewer 7: Ok, then was just a miscommunication, thanks for clarifying patrick, I was scared we were creating this tight coupling :)

  - [Failure mode brief description]
    - Detection: How can it be detected via metrics? Stated another way:
      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->
  - [Pod pending due to extended resource backed by DRA requests no less than 128 devices]
    - Detection: inspect pod status 'Pending'
    - Mitigations: reduce the number of devices requested in one extended resource backed by DRA requests
    - Diagnostics: scheduler logs at level 5 show the reason for the scheduling failure.
    - Testing: Will be considered for beta.

###### What steps should be taken if SLOs are not being met to determine the problem?

Will be considered for beta.

## Implementation History

- Kubernetes 1.34: KEP accepted.

## Drawbacks

It adds complexity to the scheduler.

> ## Alternatives

Reviewer 7: Option 3: Dra driver implement the extended resources too?

Reviewer 7: it looks way simpler of the dra driver implements the extended resources, and announces both via dra and the extended plugin

Author: we discussed this option in the early iteration of this KEP, we felt it's better to keep the change in k/k than on the driver, so that it is implemented once in k/k, rather than once per driver, which is more work on the driver developer.

Reviewer 2: If a DRA driver is also an Device Plugin, then to the scheduler the same device is advertised twice and might get allocated for one pod via extended resources and another pod via ResourceClaim -> racy, :boom:

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

Extended resource provides a simple, concise approach to describe resource
capacity, and resource consumption. In constrast, Dynamic Resource
Allocation (DRA) provides more expressive, flexible, powerful, yet more
complicated, and harder to use approach.

This KEP provides a solution to enable cluster administrators and device drivers
to advertise the resources as extended resource, and enables the application
developers, and operators to continue using extended resource to request for such
resources.

## Motivation

There are three major motivations for the solution in this KEP.

* Enable the existing applications to run without modification.

* Enable application developers, operators to transition to DRA gradually at
  their own pace.

* Enable cluster administrators to transition to DRA gradually at their own pace.

> For example, the following `Deployment` can be installed without modification on a

Reviewer 8: I assume this statement is the desired behavior that will be possible when this feature is added? The language makes it a bit unclear and the `ResourceSlice` here doesn't contain the `ExtendedResourceName` field that would be needed.

  selector:
    matchLabels:
      app: demo
  template:
    metadata:
      labels:
        app: demo
    spec:
      containers:
      - name: demo
        image: nvidia/cuda:8.0-runtime
        command: ["/bin/sh", "-c"]
        args: ["nvidia-smi && tail -f /dev/null"]
        resources:
          limits:
            nvidia.com/gpu: 1
```

```yaml
apiVersion: resource.k8s.io/v1beta1
kind: ResourceSlice
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2-gpu.nvidia.coqj92d
spec:
  devices:
  - basic:
    name: gpu-0
  - basic:
    name: gpu-1
  - basic:
    name: gpu-2
  - basic:
    name: gpu-3
  - basic:
    name: gpu-4
  - basic:
    name: gpu-5
  - basic:
    name: gpu-6
  - basic:
    name: gpu-7
  driver: gpu.nvidia.com
  nodeName: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2
```

With this motivating example in mind. We define the following goals and
non-goals of this KEP.

### Goals

* Introduce the ability for DRA to advertise extended resources listed in a
  ResourceSlice, and for the scheduler to consider them for allocation.

* Enable application operators to use the existing extended resource limits to
request for DRA resources.

### Non-Goals

> * Extended resource support is not added just for easing the transition to DRA

Reviewer 8: This sounds more like an additional goal than a non-goal.

Author: right, moved it to Goals, and added a Non-Goals.

  for the short term. Its ease of use is one big advantage to keep it remaining
  useful for the long term.

## Proposal

> The basic idea is the following:

Reviewer 9: I have been thinking about this a bit more. I think the essential idea here is to have the scheduler account for the resources in *both* the traditional extended resources AND in the DRA resources. So, the the scheduler needs to know how to between extended resource strings and DRA devices. The questions that come in are:

1) Where do we store that mapping? It could be its own resource or configuration stanza, the DeviceClass, or the ResourceSlice. The current proposal is in the ResourceSlice.
2) Where do we store the allocation? The current proposal has a special singleton ResourceClaim in a namespace for tracking this for all Pods. As @mortent pointed out, that may not work because drivers need to map between the allocations and the specific `requests` values, so that they can place devices in the correct containers. Instead, we may need to generate a per-Pod ResourceClaim, just like we do if there is a ResourceClaimTemplate for the Pod.
3) If we generate a per-Pod ResourceClaim, *when* do we do that? With a webook as described in the alternatives? Or with the ResourceClaim controller as described in the current proposal and as is done for ResourceClaimTemplates?
4) If we generate a per-Pod ResourceClaim in the ResourceClaim controller, how do we know in the scheduler to defer scheduling until that is created? I am guessing today we know if there is a ResourceClaimTemplate reference from the PodSpec, we have to wait for it to exist in order to schedule. But in this case for these *implicit* templates that exist because requests were made via the extended resources, we would have to have the same logic in the scheduler plugin to "know" that it should wait for the RC generation.
5) How does all of this work if there are existing nodes with the device plugin running that provide that extended resource? If we generate a per-pod RC, then ONLY the DRA-based drivers will be able to handle the request. Is that OK?

Reviewer 9: I think we do have enough information, if we add DeviceClass somewhere. What we need for each request is: the device class, the container, the count. Since the resource request is in the container already, we know that.  The count is also there already. It's the mapping to device class that we need to add somewhere.

In the current proposal, the mapping is between devices in the ResourceSlice and the extended resource names. We would need to add DeviceClass. However, if possible, it would be nice to avoid having to have a separate device class for every extended resource; the original "put the mapping in deviceclass" idea has that problem. But in this proposal, we could maybe allow the ResourceSlice to specify one of the following for each mapped resource name:
- deviceclassname
- deviceclassname + cel expression

I think this would allow the accounting to be done per extended resource name as this KEP suggests, we can still allow multiple extended resources to map to the same device class + a CEL selector expression.

Note that there is no reason this need apply ONLY to extended resources...we could map any resource name this way, which provides a path toward managing standard resources.

Reviewer 2: Make DeviceClassName optional?

Sorry, couldn't resist.

Reviewer 9: @pohly I think your suggestion is somewhere between what @yliaog originally suggested (no requests, directly setting allocations in the status), and what Morten & I were discussing (full claim creation). I think it could work, if we include the container name. But wouldn't that require changes to drivers to understand how to map these things to containers (right now it is via request names, I believe)? With the full claim creation, it's transparent to the driver, isn't it?

I think you are trying to address the case where we have Device Plugin running on Node A, DRA on Node B. I agree it's not clear how we address that in the other discussions so far. The discussions so far are centered around a node advertising BOTH extended resources and DRA devices via the same DRA driver, and having the scheduler account for the allocation in both places. I think the eventual solution will need to handle clusters that have some nodes running the old Device Plugin, and some nodes running the new DRA driver (whether that advertises only DRA resources or both DRA and extended resources is something we need to figure out here).

Author: Thanks all for the comments. I agree it is better to have the special claim as close to normal as possible. I revised the KEP to clarify that it supports a cluster with one node installed DRA driver, and the other node installed device plugin, and both advertised same name. I have not thought about how to support configs, my current thinking is config should be done through the regular DRA, the extended resource is a shorthand for a pool of devices, so keep it that simple way.

Reviewer 10: I think the fundamental issue being discussed here would be resolved with my proposal above. I.e. enable this feature by introducing a API server object that provides a mapping between an extended resource name and a deviceClass. Using this, it's obvious how to construct a one-off resource claim for the device.

Author: where that mapping is added (slice, device class, or a new api object) can be discussed, but that only solves one problem, i.e. how to map device to 'extended resource name'.

the other problems still exist, and need to to resolved, like
1/ where to store the allocation results (a special claim)
2/ who creates that special claim (scheduler)
3/ what is the special claim's scope, per pod per extended resource?
4/ how does scheduler noderesource plugin allow the node with the DRA backed device to fit the pod
5/ how does scheduler dynamic resource plugin do the proper accouting and allocation
6/ when (at what scheduling framework phase) is the claim created?
7/ how to link the pod to the special claim
8/ how does kubelet admit the pod to the node (its admission logic has to change)
9/ how does kubelet pass the devices to the containers inside the pod

Reviewer 9: I think the current proposal addresses all of this. Thanks!

1. Introduce a field `ExtendedResourceName` to `ResourceSlice` and `Device` to
   allow cluster administrators to configure DRA device driver to advertise
   certain devices as extended resource.
1. Introduce a special `ResourceClaim` to keep track of device allocations for
   extended resource.

With these additions in place, the scheduler has everything it needs to support
the dynamic allocation of devices to requests made through extended resource.
kubelet and DRA driver has everything it needs to admit and pass the allocated
devices to the pod to run.

## Design Details

### Resource Slice API
The exact set of proposed API changes on Resource Slice can be seen below:
```go
// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
type ResourceSliceSpec struct {
	...

> 	// The extended resource name for all the devices in the ResourceSlice

Reviewer 8: It would be helpful to include in the comments that the field on the device will override the name on the resuorceslice.

Author: done

// +k8s:deepcopy-gen=true
type Device struct {
	// Name is unique identifier among all devices managed by
	// the driver in the pool. It must be a DNS label.
	//
	// +required
	Name string `json:"name"`
	...

	// ExtendedResourceName is the extended resource name
	// the device is advertised as. It must be a DNS label.
	//
	// +optional
	ExtendedResourceName string
}
```

The devices can be advertised with an extended resource name. The extended
resource name can be specified on each individual device. Different
devices can be advertised as different extended resource name, or not
advertised as extended resource at all.

Alternatively, the extended resource name can be specified at the
`ResourceSlice` level, then all the devices in the resource slice are
advertised as the given extended resource name. If a device has a different
extended resource name than that given in the `ResoureSlice`, the device's
extended resource name is used for that device.

### Resource Claim API
There is no API change on Resource Claim. However, a special resource claim is
created to keep track of device allocations for extended resource. The special
resource claim has following properties:

>   * It is namespace scoped, like other resource claims. Unlike resource slice,

Reviewer 8: Just to clarify here, we are not talking about creating a new type right? This is just a special `ResourceClaim` in the sense that it is managed by the resource claim controller and lacks certain fields that would otherwise be expected on a ResourceClaim. Might be worth clarifying since this sentence about the scope of the resource can be a bit confusing.

Author: clarified in the KEP

    which is cluster scoped.
  * It is a singleton. There is at most one resource claim object for a given
    extended resource in a given namespace.
  * It is not owned by a pod, its owner reference is null.

>   * Its field `status.allocation.devices` is used, other fields are ununsed,

Reviewer 8: The `DeviceRequestAllocationResult` has a `Request` field that is a reference to the request in the `ResourceClaim` spec. But in this situation, there is no corresponding request in the spec (since the spec is empty). Drivers might use this value, among other things for looking up any device configuration. How do we plan to handle this?

Author: DRA driver has two parts: 
1/ 1st part that handles DRA resource claim actuation
2/ 2nd part that handles extended resources actuation

the 1st part does not need to actuate on this special resource claim object, it is ignored there.

the 2nd part needs to read the list of allocated devices from the special resource claim object, and only use the devices in that list for extended resources actuation.

Reviewer 2: This pushes the responsibility of choosing a device back into the DRA driver. I don't think this is the right direction from an architectural perspective.

cc @klueska 

I also very much dislike that DRA drivers need to be updated at all.

Reviewer 9: No, I think you misunderstood. The allocations are made by the scheduler and stored in the special singleton resource claim, not decided by the driver.

Author: DRA driver has to be updated to advertise devices as 'extended resource'. 

That said, I revised the design based on the feedback, thanks for the comments!

A cluster node install either a DRA driver, or a deivce plugin driver for a given named resource.  Devices are picked at the scheudling time, and communicated to kubelet and DRA driver through the special resource claim.

Reviewer 2: > No, I think you misunderstood (or I did). As I read it, the allocations are made by the scheduler and stored in the special singleton resource claim, not decided by the driver.

That's a bit different from what I understood. If it's still the scheduler which picks devices and tells the driver about it, then it's fine.

From https://github.com/kubernetes/enhancements/pull/5136#pullrequestreview-2607129069:
> When you and I talked on Friday, we discussed allowing kubelet to remain unchanged. Instead, it would call the Device Plugin grpc API for those with extended resources, but since that grpc API would be handled by the DRA driver, it would know when receiving calls on that API, it should look for the special resource claim. 

That probably goes back to my comment above: "I also very much dislike that DRA drivers need to be updated at all."

Why should we put additional work on all DRA drivers, now and in the future, if we can instead do something once in the kubelet? It has implications for the graduation of this feature, but should that be a deciding factor?

I know that I've said that we want to keep the kubelet as dumb as possible, but in this case I think it's simpler overall to do this in the kubelet. I'm also a bit worried about the implications of skipping some of the usual admission checks that the kubelet does for claims, like "reserved for". Perhaps that doesn't matter, but then we need to explain in the KEP why.

Author: My initial thoughts were to minimize kubelet changes, and shift the work to device driver. But after some more thoughts, I agree it's better to do the work once, and great in kubelet, so device driver can have less work to do, and less chances to implement it wrongly.

    such as `status.allocation.reservedFor` and `spec`.
  * It has annotation `resource.kubernetes.io/extended-resource-name:`, and it
    does not have annotation `resource.kubernetes.io/pod-claim-name:`

```yaml
metadata:
  annotations:
    resource.kubernetes.io/extended-resource-name: foo.domain/bar
```

The special resource claim lifecycle is managed by the resource claim
controller.

>   * It is *created* in a namespace when there is a pod with extended resource

Reviewer 8: And I assume it gets created in the same namespace as the pod?

Author: Yes, it is in the same namespace. clarified it in the KEP.

  requests, and the extended resource is advertised by Resource Slice. For
  example, A resource slice has `ExtendedResourceName: foo-domain/bar`, and a
  pod requests for foo-domain/bar.
  * It is *deleted* with the namespace deletion. (this is simpler, since there
  is only one such resource claim object per namespace per extended resource,
  it won’t cause much extra overhead to leave it around)

>   * Alternatively, it is *deleted* by the resource claim controller when there

Reviewer 8: So this sentence and the previous one means that the `ResourceClaim` will get deleted when the Pod is deleted. But we just don't have any automatic collection using the owner refs?

Author: correct. either namespace controller, or the resource claim controller will delete the special resource claim object when it is no longer needed, I.e. when there is no pod in the namespace that has request for that extended resource.

Author: this has been revised, the claim will have pod as owner, so that it is as close to regular claim as possible.

          limits:
            nvidia.com/gpu: 1
```

```yaml
apiVersion: resource.k8s.io/v1beta1
kind: ResourceSlice
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2-gpu.nvidia.coqj92d
spec:
  devices:
  - basic:
    name: gpu-0
  - basic:
    name: gpu-1
  - basic:
    name: gpu-2
  - basic:
    name: gpu-3
  - basic:
    name: gpu-4
  - basic:
    name: gpu-5
  - basic:
    name: gpu-6
  - basic:
    name: gpu-7
  driver: gpu.nvidia.com
  nodeName: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2
  extendedResourceName: nvidia.com/gpu
```

```yaml
apiVersion: v1
kind: Node
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2
status:
  allocatable:
    cpu: 3920m
    ephemeral-storage: "49455357798"
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 12636272Ki
    pods: "110"
    nvidia.com/gpu: "8"
  capacity:
    cpu: "4"
    ephemeral-storage: 101430960Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 15335536Ki
    pods: "110"
    nvidia.com/gpu: "8"
```

With this motivating example in mind. We define the following goals and
non-goals of this KEP.

### Goals

* Introduce the ability for DRA to advertise extended resources listed in a

>   ResourceSlice, and for the scheduler to consider them for allocation.

Author: I mean the former, not the latter. I clarified that in the KEP. an extended resource is either advertised in ResourceSlice (by DRA), or in Node's status.Capacity(by device plugin).

* Enable application operators to use the existing extended resource limits to

> request for DRA resources.

Author: rephrased, hopefully it is clear.

	// the device is advertised as. It must be a DNS label.
	// It overrides the ExtendedResourceName at ResourceSlice if both are
	// present.
	//
	// +optional
	ExtendedResourceName string
}
```

The devices can be advertised with an extended resource name. The extended
resource name can be specified on each individual device. Different
devices can be advertised as different extended resource name, or not
advertised as extended resource at all.

Alternatively, the extended resource name can be specified at the
`ResourceSlice` level, then all the devices in the resource slice are
advertised as the given extended resource name. If a device has a different
extended resource name than that given in the `ResoureSlice`, the device's
extended resource name is used for that device.

### Resource Claim API
There is no API change on `Resource Claim`, i.e. no new API type. However, a special
resource claim object is created to keep track of device allocations for extended
resource. The special resource claim object has following properties:

  * It is namespace scoped, like other resource claim objects.
  * It is a singleton per node. There is at most one such resource claim object for

>     a given extended resource in a given namespace per node that advertises the

Author: for the risky comment, it is the same as other claims in terms of permission. that said, i agree it is better to keep it as close to regular claim as possible. The KEP is revised to reflect that.

allocation for future requests from either extended resource, or DRA resource claim.

Note that another node in the same cluster has installed device plugin, which
has advertised 'nvidia.com/gpu: 2' in its `Node`'s Capacity. The same
`Deployment` can possibly be scheduled and run on this node too.

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: demo
spec:
  replicas: 1
  selector:
    matchLabels:
      app: demo
  template:
    metadata:
      labels:
        app: demo
    spec:
      containers:
      - name: demo
        image: nvidia/cuda:8.0-runtime
        command: ["/bin/sh", "-c"]
        args: ["nvidia-smi && tail -f /dev/null"]
        resources:
          limits:
            nvidia.com/gpu: 1
```

```yaml
apiVersion: resource.k8s.io/v1beta1
kind: ResourceSlice
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2-gpu.nvidia.coqj92d
spec:
  devices:
  - basic:
    name: gpu-0
  - basic:
    name: gpu-1
  - basic:
    name: gpu-2
  - basic:
    name: gpu-3
  - basic:
    name: gpu-4
  - basic:
    name: gpu-5
  - basic:
    name: gpu-6
  - basic:
    name: gpu-7
  driver: gpu.nvidia.com
  nodeName: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2

>   extendedResourceName: nvidia.com/gpu

Reviewer 2: Agreed, see my comment on the API in https://github.com/kubernetes/enhancements/pull/5136#discussion_r1949136336

>   extendedResourceName: nvidia.com/gpu

Author: Agreed, we also have the per-device  `extendedResourceName` in this KEP, so the question is whether we need this slice-scope setting.

>   extendedResourceName: nvidia.com/gpu

Author: we decide to use deviceclass for extended resource -> device mapping, updated that in the KEP.

```

```yaml
apiVersion: v1
kind: Node
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2
status:
  capacity:
    cpu: "4"
    ephemeral-storage: 101430960Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 15335536Ki
    pods: "110"
```

```yaml
apiVersion: v1
kind: Node
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-xyz8
status:
  capacity:
    cpu: "4"
    ephemeral-storage: 101430960Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 15335536Ki
    pods: "110"
    nvidia.com/gpu: 2
```

With this motivating example in mind. We define the following goals and
non-goals of this KEP.

### Goals

> * Introduce the ability for DRA to advertise extended resources listed in a

Author: Yes, admin should be able to configure the mapping in ResourceSlices.

> * Introduce the ability for DRA to advertise extended resources listed in a

Reviewer 9: Admins generally don't touch ResourceSlices directly, they are created and managed by the DRA drivers. So, if we make this a *field*, then admins really won't be able to touch it. @pohly is saying that if it an *attribute*, then #5027 can be used to let the admin tweak the mapping.

> * Introduce the ability for DRA to advertise extended resources listed in a

Author: ok, i have not read #5027, i could be missing the point. but i think admin should be able to configure device driver to advertise a device.

> * Introduce the ability for DRA to advertise extended resources listed in a

Reviewer 2: > I am strongly against allowing the admin to define / set this directly in the resource slice. If we want admins to be able to control this mapping, then it should be done in a separate API server object, not the resourceSlice itself.

Agreed, and no-one is suggesting that. We have two options:
1. An admin reconfigures the DRA driver so that it includes this information ("configure device driver to advertise a device").
1. An admin uses the ResourceSlicePatch object from https://github.com/kubernetes/enhancements/issues/5027 to add the special attribute. This does not change the ResourceSlice, only how the scheduler treats it.

> My original idea when I created the issue for this feature was to introduce a new API server object which provided a mapping between a deviceClass and an extended resource name.

Or even better: make this a new list field in DeviceClass. Then we don't need a separate mapping object. My initial concern was about granularity, but the DeviceClass can be defined such that it is as specific as it needs to be to match the semantic of the extended resource name(s).

We could theoretically then have two DeviceClasses that both map to the same extended resource name. The semantic of that would be similar to the `firstAvailable`: pick a device matching the first class, otherwise the second. Once we have scoring, we can make that "pick the best device which matches one of these two class".

> * Introduce the ability for DRA to advertise extended resources listed in a

Author: so i see three levels of granularity: 1/ device level 2/ device class level 3/ slice level (4/ as klueska suggested a new api object for mapping device class to extended resource name, which is still the device class level)

so the first question to ask is: which level of granularity is the best? the answer very much depends on the use cases we need to support.

and the second question is centered around the usability/complexity, i can see the extra flexibility / expressiveness suggested above, my question is: are they needed? are they common use cases?

"device-level, and an admin reconfigures the DRA driver to advertise a device" proposed in this KEP is close to how the current existing extended resource is configured, i know we don't have to stick to it.

  ResourceSlice, and for the scheduler to consider them for allocation.

* Enable application operators to use the existing extended resource request in
  pod spec to request for DRA resources.

* Extended resource support is not added just for easing the transition to DRA
  for the short term. Its ease of use is one big advantage to keep it remaining
  useful for the long term.

### Non-Goals

* Simplify DRA driver developement. The DRA driver needs to support both DRA

>   and extended resource API. This KEP adds complexity and cost to the driver.

Reviewer 2: > Is this saying that (as a DRA driver deveoper), I need to effectively "merge" my DRA driver with the existing device plugin logic and allow allocation requests to be honored via either API?

That was the original proposal from last week before I got involved. This non-goal here is a left-over from that, it no longer matches the current KEP text, hence my suggestion to change it.

You and I both agree that keeping DRA driver development simple should be a goal.

>   and extended resource API. This KEP adds complexity and cost to the driver.

Reviewer 2: See also https://github.com/kubernetes/enhancements/pull/5136#pullrequestreview-2607129069 from @johnbelamaric 

Besides the complexity, I'm not even sure whether it would work in all cases: when a DRA driver pretends to be a device plugin by implementing the device plugin gRPC interface, it cannot rely in kubelet to inject CDI devices into the containers, instead having to use some other API. Or can the driver inject a CDI device itself? Not sure.

Either way, let's not do this.

>   and extended resource API. This KEP adds complexity and cost to the driver.

Reviewer: @pohly 
> when a DRA driver pretends to be a device plugin by implementing the device plugin gRPC interface, it cannot rely in kubelet to inject CDI devices into the containers

I understand that we're not going to do that, just want to point out that [Device Plugins API supports CDI devices](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4009-add-cdi-devices-to-device-plugin-api)

>   and extended resource API. This KEP adds complexity and cost to the driver.

Author: Yes, we are all on the same page on this, i.e., reduce the work on device driver, shift it to kubelet. I'll update the KEP to avoid the confusion around this.

>   and extended resource API. This KEP adds complexity and cost to the driver.

Reviewer: @yliaog 
> reduce the work on device driver, shift it to kubelet

According to [this comment](https://github.com/kubernetes/enhancements/pull/5136#discussion_r1950562400) the most of work should be done by control plane, not by Kubelet.

>   and extended resource API. This KEP adds complexity and cost to the driver.

Reviewer 2: Which may not be feasible (https://github.com/kubernetes/enhancements/pull/5136#discussion_r1952316855).

>   and extended resource API. This KEP adds complexity and cost to the driver.

Author: @bart0sh , scheduler does a lot of heavy lifting work for sure, but kublet has to had some change, not that much compared to that for scheduler.

## Proposal

The basic idea is the following:

1. Introduce a new concept `dynamic extended resource`. It is like the current
   extended resource, in that, it has a string name, and a discrete countable
   quantity. Its capacity is provided through dynamic resource `ResourceSlice`,

>    its consumption is specified through current pod's extended resource request.

Author: we need to have a way to distinguish the extended resource backed by DRA, vs the extended resource backed by k8s Node (device plugin), this KEP needs to apply different scheduling logic to them, and also different actuation logic in kubelet.

>    its consumption is specified through current pod's extended resource request.

Reviewer 10: Maybe I'm missing some fundamental assumtion here -- but why must the actuation logic in the kubelet be different.

I would expect the scheduler to:
1. Intersect the extended resource request
2. Gather the set of nodes where it is being advertised via the traditional device plugin API
3. Check to see if there is a mapping set up for the extended resource to a specific device class and gather the set of nodes where that deviceClass can be satisfied
4. Evaluate between (2) and (3) where the best place to allocate the resource is
5. If (2) is chose, trigger the standard device plugin path
6. If (3) is chose, create a one-off resource claim for the device, inject it into the pod and trigger the DRA path

>    its consumption is specified through current pod's extended resource request.

Author: At the very high level, what you described is the steps the scheduler takes, this KEP provides more concrete design on how the steps can be fitted into the current scheduler framework.

back to the questiion in this comment thread, scheduler needs to treat the extended resource backed by plugin API *different* from that backed by DRA API, this is clearly shown in the 6 steps you listed.

>    its consumption is specified through current pod's extended resource request.

Reviewer 10: I think it should be an explicit goal that *no* changes are required in:
1. The DRA driver API (including the devices advertised in its ResourceSlice)
2. The device plugin API
3. The kubelet

All changes should be isolated to the scheduler and any new or existing abstractions that it operates on.

>    its consumption is specified through current pod's extended resource request.

Author: 1. The DRA driver API (including the devices advertised in its ResourceSlice): 
This needs further discussion, there are a few options on the table, we need to discuss which is the best from felxibility, and usability perspective.

2. The device plugin API
For sure, there must be no change

3. kubelet
there must be changes to its admission logic to allow the pod with extended resource backed by DRA to run on the node
there must be changes to pass DRA prepared devices into container's with extended resource requests backed by DRA

>    its consumption is specified through current pod's extended resource request.

Reviewer 9: The only way to avoid kubelet changes is with an admission webhook that converts extended resource requests into resource claims and updates the podspec to point to those claims. That doesn't sound ideal to me.

>    its consumption is specified through current pod's extended resource request.

Reviewer 10: Why can't the scheduler do this?

>    its consumption is specified through current pod's extended resource request.

Author: after scheduler has scheduled a pod to a node, and the node kubelet picks up the pod, kubelet also has the admission logic to check if the pod can actaully run on the node, that check needs changes to allow 'extended resource requests' in the pod to fit on the node with devices backed by DRA.

>    its consumption is specified through current pod's extended resource request.

Reviewer 10: > after scheduler has scheduled a pod to a node, and the node kubelet picks up the pod, kubelet also has the, admission logic to check if the pod can actaully run on the node, that check needs changes to allow 'extended resource requests' in the pod to fit on the node with devices backed by DRA.

I realize this is your proposal, but I'm questioning whether we actually need / want this. 

My comment was meant to ask why the scheduler can't just modify the pod -- or more specifically some plugin in the scheduler that sits in front of its logic to evaluate either extended resources or resource claims.

I realize this is not something done today, but it's not clear to me why we can't make it do this in the future to enable this feature.

>    its consumption is specified through current pod's extended resource request.

Reviewer 2: I don't think we can have the scheduler rewrite the pod spec. This will very likely confuse app controllers which don't expect such a modification.

There are a few fields that the scheduler is allowed to change, most notably the `nodeName` for recording the binding to a node. But these are well-known exceptions.

>    its consumption is specified through current pod's extended resource request.

Reviewer 10: If that is our starting point then so be it.

As mentioned in the KEP, it then forces *some* branching mechanism to exist in the `kubelet` to decide whether to service an incoming extended resource as a DRA allocated resource or a standard device plugin resource.

I'll need to take a closer look at the rest of this KEP ow to see what is being proposed to enable this.

That said, I'm still skeptical that we need (or even want to!) add a field to the ResourceSlice (or any of its devices) specify that those devices are allocatable via an extended resource name. This seems better suited as a field in a DeviceClass (or as a new API server object linking a DeviceClass to an extended resource type).

>    its consumption is specified through current pod's extended resource request.

Reviewer 2: I myself prefer the approach with a new field in the DeviceClass.

>    its consumption is specified through current pod's extended resource request.

Reviewer 10: With something like that in place, I could imagine a flow like the following for servicing `nvidia.com/gpu` devices:

1. An admin creates a DeviceClass mapping (select) devices from the `gpu.nvidia.com` DRA driver to the extended resource name `nvidia.com/gpu`
2. The user submits a request `nvidia.com/gpu: 2`
3. The scheduler decides to allocate these 2 GPUs on a node via the DeviceClass providing this mapping
4. The scheduler creates a one-off resourceClaim requesting 2 GPUs from that DeviceClass and allocates 2 GPUs immediately to it.
5. The scheduler puts a mapping from pod/container --> this one-off resourceClaim in the status of the DeviceClass
6. The scheduler schedules the pod to the node
8. When the kublet picks up the pod with a container asking for `nvidia.com/gpu: 2` it checks to see whether it should service this via the standard device plugin or DRA. It does this by first looking to see if there is a standard device plugin registered with `nvidia.com/gpu`. If there is, it uses that plugin to service the request. If there isn't, it checks to see if there is a DeviceClass with a mapping for `nvidia.com/gpu` and that the driver providing this mapping is registered with the kubelet, as well as a mapping from pod/container --> resourceClaim in the DeviceClass's status for the pod/container under question.
9. Once decided, it goes down one of these two paths to service the request.
10. Having both a device plugin and a DRA driver registered on the node servicing the same type of extended resource should be considered a misconfiguration.

>    its consumption is specified through current pod's extended resource request.

Author: @klueska thanks for the detailed steps listed above. it is mostly aligned what we proposed in this KEP, a few differences listed below:

1/ pod/container --> this one-off resourceClaim is recorded in pod.status.extendedResourceClaimStatuses, similar to how other regular resource claim is recorded in pod.status.resourceClaimStatuses
2/ device class has only one optional field 'extendedResourceName' to indicate the devices selected by this devicelass has a name that can be requested through nvidia.com/gpu: 2 in pod.spec.resources.request
3/ in kubelet, it gets the special resourceclaim from pod.status.extendedResourceClaimStatues in the same way it gets other regular resourceclaims from pod.status.resourceClaimstatues, then kubelet pass the claim IDs to DRA driver, DRA driver the calls k8s API GET resourceclaim, then from there to retrieve the list of devices, and then DRA driver prepare the devices, and return the CDI device ids back to kubelet.
4/ with the CDI device IDs, kubelet insert them into the container that requests "nvidia.com/gpu", then pass it to container runtime to start the container, with the devices.
5/ Yes about 9. above, a node can only have either one extened resource name backed by device plugin, or by DRA, not both.

>    its consumption is specified through current pod's extended resource request.

Author: regarding the earlier question about mutating pod spec in scheduler, like @pohly mentioned, except few exceptions like the pod.spec.nodeName, pod spec is immutable after pod admission into the k8s API server, pod spec is mutable at the pod admission time with the webhook mechanism mentioned by @johnbelamaric . 

this has been true for the past 10 years of k8s. that said, there is a new proposal for in-place pod update, that is still being worked on though.

1. Introduce a field `ExtendedResourceName` to `ResourceSlice` and `Device` to
   allow cluster administrators to configure DRA device driver to advertise
   certain devices as extended resource. This is different from the extended
   resource advertised by k8s `Node` Object, which are usually backed by device

>    plugin. Hence they are given a new name 'dynamic extended resource'.

Author: updated the KEP.

1. Introduce a special `ResourceClaim` object to keep track of device allocations

>    for each dynamic extended resource requests for a pod.

Author: done

   `ResorceSlice`.
1. kubelet asks DRA driver to prepare devices in the special `ResourceClaim`,
   and pass the devices to containers with the dynamic extended resource
   requests.

Some quick clarifications around the basic concepts: extended resource, dynamic
resource, and dynamic extended resource.

* extended resource uses pod's spec.containers[].resources.requests to request
  for resources, extended resource consumes the capacity from node's
  status.capacity. exteneded resource is of type: string, int
* dynamic resource uses `ResourceClaim` to request for resources, and
  `ResourceSlice` to provide resource capacity. A pod asks for resources through
  resource claim requests in pod's spec.resources.claims. dynamic resource type
  is described in resource slice, simply speaking, it is a list of devices, with
  each device being described as structured paramaters.
* dynamic extended resource is a combination of the two above. It uses pods'
  spec.containers[].resources.requests to request for resources, and uses
  `ResourceSlice` to provide resource capacity. Hence, it is of type: string,
  int on the consumption side, and list of devices with a common

>   `ExtendedResourceName` on the capacity side.

Author: this KEP needs to distinguish how a request gets satisfied, as it requires different code logic in the scheduler, and kubelet, even though the user of the request does not care.

>   `ExtendedResourceName` on the capacity side.

Reviewer 2: My point still stands: We don't need to invent a new name for this.

>   `ExtendedResourceName` on the capacity side.

Author: ok, how about this, we use the term "extended resource backed by device plugin" and "extended resource backed by DRA" to distinguish two in this KEP?

>   `ExtendedResourceName` on the capacity side.

Reviewer 2: That makes sense to me.

>   `ExtendedResourceName` on the capacity side.

Author: updated the KEP to use  the term "extended resource backed by device plugin" and "extended resource backed by DRA"

requests, or by DRA resouce claims. The scheduler has everything it needs to support
the dynamic allocation of devices to requests made through extended resource and
resource claims. No static partition of resources between extended resources and
resource claims is needed. The kubelet and DRA driver has everything they need
to admit and pass the allocated devices to the pod to run.

The following cluster setup is supported.

* One node in cluster has a dynamic extended resource, and another node in the
cluster has the same name extended resource.

* One node in a cluster cannot have both dynamic extended resource, and same
  name extended resource at the same time. This implies that either the resource
  is advertised through `ResourceSlice`, or `Node`'s status.capacity.

## Design Details

### Resource Slice API
The exact set of proposed API changes on Resource Slice can be seen below:
```go
// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
type ResourceSliceSpec struct {
	...

	// The extended resource name for all the devices in the ResourceSlice
	// advertised as
	//
	// +optional

> 	ExtendedResourceName string

Author: Got the point about using `gpu.example.com`.
For #5034 using ResourceSlicePatch, we need further discussion to decide which option is the best.

}

// Device represents one individual hardware instance that can be selected based
// on its attributes. Besides the name, exactly one field must be set.
// +k8s:deepcopy-gen=true
type Device struct {
	// Name is unique identifier among all devices managed by
	// the driver in the pool. It must be a DNS label.
	//
	// +required
	Name string `json:"name"`
	...

	// ExtendedResourceName is the extended resource name
	// the device is advertised as. It must be a DNS label.
	// It overrides the ExtendedResourceName at ResourceSlice if both are
	// present.
	//
	// +optional

> 	ExtendedResourceName string

Author: ok. we can do just per-device level, without slice level mapping.

}
```

The devices can be advertised with an extended resource name. The extended
resource name can be specified on each individual device. Different
devices can be advertised as different extended resource name, or not
advertised as extended resource at all.

Alternatively, the extended resource name can be specified at the
`ResourceSlice` level, then all the devices in the resource slice are
advertised as the given extended resource name. If a device has a different
extended resource name than that given in the `ResoureSlice`, the device's

> extended resource name is used for that device.

Author: a string-list seems to be future proof in case there is a use case for a device to have multiple extended resource names.

> extended resource name is used for that device.

Reviewer 2: This can be avoided by doing the mapping in the DeviceClass, see https://github.com/kubernetes/enhancements/pull/5136#discussion_r1950622581.

### Resource Claim API
There is no API change on `ResourceClaim`, i.e. no new API type. However, a special
resource claim object is created to keep track of device allocations for dyanmic
extended resource. The special resource claim object has following properties:

  * It is namespace scoped, like other resource claim objects.
  * It is owned by a pod, like other resource claim objects.
  * It has null `spec`.
  * Its `status.allocation.devices` and `status.allocation.reservedFor` are
    used.
  * It has annotation `resource.kubernetes.io/extended-resource-name:`, and it
    does not have annotation `resource.kubernetes.io/pod-claim-name:`

```yaml
metadata:
  annotations:
    resource.kubernetes.io/extended-resource-name: foo.domain/bar

> ```

Reviewer 2: At this point, we have avoided adding new special annotations and new fields. Relaxing validation of `pod.spec.resourceClaims[].name` is not good as it breaks downgrades!

A cleaner solution is:
* `resource.kubernetes.io/extended-resource-claim: True` for the claim annotation
* a separate `DeviceAllocationResult.ExtendedResults`
* a separate `PodStatus.ExtendedResourceClaimStatuses`

My first thought was that we could add a `DeviceRequestAllocationResult.Container` field as one-of with `Request`, but that also breaks downgrades: the previous release will see an empty `Request`, which was invalid.

> ```

Author: Quoted From KEP 4381 below, so it says claim.requests can be null, does that mean we can leave DeviceRequestAllocationResult.Request null too?

=============================
"There are some corner cases:

Empty claim.requests: this is a "null request" which can be satisfied without allocating any device.
Non-empty deviceClassName, empty class: this is how admins can enable users to select arbitrary devices."
=============================

> ```

Author: foo.example.com/gpu and bar.example.com/gpu are two different extended resource names, they should be allocated with two claims, that is cleaner, and more straightforward, IMO.

> ```

Reviewer 2: > does that mean we can leave DeviceRequestAllocationResult.Request null too?

What we would leave empty is `DeviceAllocationResult.Results`. That is valid in current Kubernetes.

> ```

Author: updated the KEP. DeviceExtendedResourceRequestAllocationResult has a field 'ExtendedResourceName', so one pod's requests for extended resources backed by DRA foo.example.com/gpu and bar.example.com/gpu are recorded in one claim, in two results.

The special resource claim object lifecycle is managed by the scheduler and
garbage collector.

  * It is *created* in a namespace when there is a pod with extended resource
  request, and the extended resource is advertised by `ResourceSlice` and
  scheduler has fit the pod to a node with the `ResourceSlice`.
  * It is *created* by the scheduler dynamic extended resource plugin during
    pre-bind phase. The in-memory one in the assumed cache is created earlier
    during reserve phase.
  * It is *deleted* together with the owning pod's deletion.
  * It is *read* by scheduler dyanmic resource plugin for the devices allocated,
  so that the scheduler remove considerations for allocation of these devices for
  other DRA resource claim requests in 'dynamic resource plugin'.
  * It is *read* by the kubelet DRA device driver to prepare the devices listed
    therein when preparing to run the pod.

### Pod API

> There is no API change on `Pod`. Pod's status.resourceClaimStatuses tracks the

Reviewer 2: There is: relaxing validation of `name` is an API change. `name: foo.domain/bar` should be invalid at the moment (not a DNS label, not the name of an entry in pod.spec.resourceClaims).

Author: as discussed in another comment, we probably don't want to relax the validation of name, to use it for extended resource name. it's cleaner to have a separate pod.status.extendedResourceClaimStatus.

special resouceclaim object created for the dynamic extended resource requests
in the pod. The dynamic extended resource name is used in the status. For
example, if a pod has requested for foo.domain/bar, and it is scheduled to run
on a node that has advertised foo.domain/bar in `ResourceSlice`, then the pod's
status is like below:

```yaml
status:
   resourceClaimStatuses:
   - name: foo.domain/bar
     resourceClaimName: ccc-gpu-57999b9c4c-vpq68-gpu-8s27z
```

### Scheduling for Dynamic Extended Resource

A new field `DynamicResources` is added to `Resource`, it works similar to
ScalarResources. It is used to keep track of the dynamic extended resources on a
node, i.e. those that are advertised by `ResourceSlice`.

```go
type Resource struct {
	MilliCPU         int64
	Memory           int64
	EphemeralStorage int64
	// We store allowedPodNumber (which is Node.Status.Allocatable.Pods().Value())
	// explicitly as int, to avoid conversions and improve performance.
	AllowedPodNumber int
	// ScalarResources
	ScalarResources map[v1.ResourceName]int64

    // NEW!
    // DynamicResources: keep track of dynamic extended resources

> 	DynamicResources map[v1.ResourceName]int64

Reviewer 2: Do we need this?

If a pod has a non-empty `PodStatus.ExtendedResourceClaimStatuses`, the kubelet knows that it needs to prepare those.

If normal extended resources were used, we can record that in `ScalarResources`, as before.

Author: This is used for scheduler, not for kubelet.

For each node, scheduler needs to know whether it has provided the capacity for the extended resource (from Node.Capacity), or for the dynamic extended resource(from ResourceSlice). For the former, it is included in the ScalarResources, for the latter, we need to add that to the DynamicResources field.

The overall flow is, at the beginining of the scheduling cycle, scheduler takes a snapshot of the nodes in hte cluster, then the noderesource plugin can check if the resource is ScalarResources or DynamicResources, and keep the node if it can satisfy the pod's extended resource request from either ScalarResources or DynamicResources. After that, the 'dynamciresource' plugin needs to actually allocate the devices for the pods's claims, and then After that, the 'dynamic extended resource' plugin needs to actually allocate the devices for the pod's dynamic extended resource. (note: the dynamic extended resource plugin may be merged into the dynamicresouce plugin as you suggested, we need to think which way is more cleaner.)

Reviewer 2: Can't the `noderesources` use the normal `ScalarResources`? That's correct for nodes where a device plugin was used and will be used again for other pods. For nodes with a DRA driver, it doesn't matter because the `noderesources` plugin should ignore the extended resource request.

I think this will become clearer once there is a prototype.

Reviewer 2: > This is used for scheduler, not for kubelet.

I didn't understand that you were talking about `k8s.io/kubernetes/pkg/scheduler/framework.Resource`. I only realized when reading https://github.com/kubernetes/enhancements/pull/5136#discussion_r1950063395.

That is the struct that you mean here, right? I thought it was an API struct. Please, don't assume that your readers can read your mind :sweat_smile: 

A link to the source code defining that struct and an explicit statement that this is about scheduler internals would be useful.

Author: Yes, that is the struct. I'll add more details to the KEP to make it clear to avoid future confusion.

}
```

type `NodeInfo` is used by scheduler to keep track of the information for each
node in memory. Its `Allocatable` field is used to keep track of the allocatable
resources in memory. For a node with extended resources, its NodeInfo's
Allocatable.ScalarResources is updated with the `Node`'s informer, minus the
used. For a node with dynamic extended resources, its NodeInfo's
Allocatable.DynamicResources is updated with the `ResourceSlice`'s informer,
minus used by either dynamic extended resource or resource claims.

The existing 'noderesources' plugin needs to be modified, such that a pod's
extended resource request is checked against a NodeInfo's ScalarResources if the
node uses device plugin, and checked against a NodeInfo's DynamicResources if
the node uses DRA driver and it advertised the extended resource in its
`ResourceSlice`.

The existing 'dynamicresources' plugin needs no modification. Since the special
resource claim for dynamic extended resource has accounted for the resource

> usage.

Reviewer 2: This is going to be problematic. Suppose a pod uses claims *and* extended resources which might be backed by DRA. Then we want to execute a single `Allocate` call to implement backtracking. It's also going to lead to a lot of code duplication.

Instead, the existing plugin could be extended to create a "fake" ResourceClaim for all extended resources that it needs to allocate. But how to determine those upfront is a bit open, same as it is for the `noderesources` plugin.

I suspect some prototyping will be needed to arrive at a solution. This makes declaring this KEP "implementable" for 1.33 a stretch: we think it can be done, but don't know yet how exactly.

I'll have to defer to SIG scheduling reviewers/approvers to provide guidance here.

Author: At the begining of the scheduling cycle, scheduler takes a snapshot of all nodes in the cluster (https://github.com/kubernetes/kubernetes/blob/80bf507facbd11711ef322f2c992c68994228785/pkg/scheduler/schedule_one.go#L413), this snapshot remains the same during the scheduling cycle.

My thinking is to add the following as soon as the snapshot is taken if DRA is enabled:
1/ take a snapshot of the resourceslices
2/ based on the resourceslices snapshot, update the NodeInfo in the node snapshot when/if the Node has Resource Slices that have advertised 'extended resources', such that, the in-memory NodeInfo has something like 'foo.domain/bar: 8', and this is tracked in NodeInfo.Resource.DynamicResources[]

With the updated NodeInfo snapshot, 
3/ noderesource plugin can check if the 'extended resource request' can be fit from a Node's ScalarResources or DynamicResources
4/ 'dynamicresource' plugin (or a separate plugin) can actually Allocate devices for the claims, and also the requests in 'extended resources'. The 'fake' ResourceClaim can be created based on the pod's 'extended resource requests' that are in NodeInfo.Resource.DynamicResources[].

Reviewer 2: Yes, something like this should work. I'm a bit worried about the overhead of collecting this per-node information over and over. It's tempting to ask for special node labels, but we can revisit this based on benchmarking some prototype.

Author: I think we are doing this kind of slice/claim snapshot inside DRA plugin already, no? IIUC, the current DRA plugin get all the slices for capacity, and all the claims for usage, and based on that, to allocate remaining available devices to pod unallocated claims.

#### EventsToRegister
This registers all cluster events that might make an unschedulable pod schedulable,
like finishing the allocation of a claim, or resource slice updates.

#### Score
If two nodes can fit the pod, one node has installed device plugin, the other
has installed DRA, score the DRA node higher.

#### Reserve
Allocate the specific devices for the number of dynamic extended resource
requested in the pod, create an in-memory `ResourceClaim` for storing the
allocation results. The claim is stored in the map of in-flight claims.

#### Prebind
This is called in a separate goroutine. The plugin makes API call to create the
`ResourceClaim` and updates the pod's status. If some API request fails now,
PreBind fails and the pod must be retried.

#### Unreserve
The plugin removes the Pod from the claim.status.reservedFor field, removes the
special `ResourceClaim` for dynamic extended resource. Because it cannot be scheduled
after all.

### Actuation for Dynamic Extended Resource
When a pod with dynamic extended resources is picked up by the kubelet on the node it
is scheduled to run, the following are particularly important:

> 1. Kubelet tries to admit the pod, the pod's dynamic extended resources requests

Reviewer 2: ```suggestion
1. Kubelet tries to admit the pod, the pod's dynamic extended resources requests
```

There wouldn't be any of those, just `ScalarResources` and claims.

Author: do you mean just avoid introducing 'dynamic extended resources' ? it's ok to avoid that. however, the admission change below is still needed.

Reviewer 2: I think we can drop `Resource.DynamicResources` and in the text stop using the term "dynamic extended resources": "extended resource request" is the existing API that we want to support, it doesn't suddenly change into something else.

Reviewer 2: Which admission change?

Reviewer 2: Oh, the admission check in the kubelet. Yes, that's still needed, depending on what was allocated.

Author: Resource.DynamicResources is introduced for scheduler, such that the plugin 'noderesource' can keep (don't filter out) the node with devices in ResourceSlice for the pod with 'extended resource request'. Agreed that on the consumption side, the request is the same as extended resource. However, on the capacity side, it is very different, in that Node.Capacity/Allocatable does not show it at all.

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: demo
spec:
  replicas: 1
  selector:
    matchLabels:
      app: demo
  template:
    metadata:
      labels:
        app: demo
    spec:
      containers:
      - name: demo
        image: nvidia/cuda:8.0-runtime
        command: ["/bin/sh", "-c"]
        args: ["nvidia-smi && tail -f /dev/null"]
        resources:
          limits:
            nvidia.com/gpu: 1
```

```yaml
apiVersion: resource.k8s.io/v1beta1
kind: ResourceSlice
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2-gpu.nvidia.coqj92d
spec:
  devices:
  - basic:
    name: gpu-0
  - basic:
    name: gpu-1
  - basic:
    name: gpu-2
  - basic:
    name: gpu-3
  - basic:
    name: gpu-4
  - basic:
    name: gpu-5
  - basic:
    name: gpu-6
  - basic:
    name: gpu-7
  driver: gpu.nvidia.com
  nodeName: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2
  extendedResourceName: nvidia.com/gpu
```

```yaml
apiVersion: v1
kind: Node
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2
status:
  capacity:
    cpu: "4"
    ephemeral-storage: 101430960Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 15335536Ki
    pods: "110"

> ```

Reviewer 10: I'm fine with that -- we should state this explicitly somewhere in the KEP though. I know alot of people rely on inspecting the Capacity/Allocatable to see whats available and might get confused if their pod lands on a node without this (when using the standard extended resource name) and is able to get hold of a device that's not listed.

> ```

Author: scheduler is the consumer of the node.status.capacity, kubelet (with the info collected from device plugin) is the producer.

scheduler checks node.status.capacity to determine whether the extended resource is provided by device plugin, and if yes, it will not check DRA ResourceSlice. If scheduler were to update the node.status.capacity, there would be a circle.

It is much simpler, and clearer to user to check where the extended resource is provisioned, either from device plugin in node.status.capacity, or from DRA in the special resource claim.

> ```

Reviewer 10: I can still see arguments both ways...

Can we add some wording in the KEP to say that in `alpha` we will not be updating these fields, but still leave it open to update them potentially in `beta`?

> ```

Author: ok, added words to the last goal in Goals section

* Keep advertising only extended resources backed by device plugin in `node.status.Capacity`
  for Alpha. It will be revisited for Beta, based on Alpha feedback.

}
```

A special resource claim object is created to keep track of device allocations for
extended resource. The resource claim object has following properties:

  * It is namespace scoped, like other resource claim objects.
  * It is owned by a pod, like other resource claim objects.
  * It has null `spec`.
  * Its `status.allocation.devices` and `status.allocation.reservedFor` are
    used.
  * It does not have annotation `resource.kubernetes.io/pod-claim-name:` as
    it is created for the extended resource request in a pod spec, not for a
    claim in the pod spec.

The special resource claim object lifecycle is managed by the scheduler and
garbage collector.

  * It is *created* in a namespace when there is a pod with extended resource
  request, and the extended resource is advertised by `ResourceSlice` and
  scheduler has fit the pod to a node with the `ResourceSlice`.
  * It is *created* by the scheduler dynamic extended resource plugin during
    preBind phase. The in-memory one in the assumed cache is created earlier
    during reserve phase.
  * It is *deleted* together with the owning pod's deletion.

>   * It is *deleted* by the scheduler dynamic extended resource plugin during

Reviewer 11: Why can't we use ownerRef and k8s gc instead?

Reviewer 2: That's the previous bullet point. This one is about the pod not proceeding to be bound: it doesn't get deleted, so the scheduler has to roll back.

Reviewer 2: In other words, we need to do both depending on the circumstances.

@yliaog: having two sub-bullets under "It is *deleted* ..." might make it clearer that these two alternative outcomes.

spec:
  devices:
  - basic:
    name: gpu-0
  - basic:
    name: gpu-1
  - basic:
    name: gpu-2
  - basic:
    name: gpu-3
  - basic:
    name: gpu-4
  - basic:
    name: gpu-5
  - basic:
    name: gpu-6
  - basic:
    name: gpu-7
  driver: gpu.example.com
  nodeName: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2
```

```yaml
apiVersion: v1
kind: Node
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2
status:
  capacity:
    cpu: "4"
    ephemeral-storage: 101430960Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 15335536Ki
    pods: "110"
```

```yaml
apiVersion: v1
kind: Node
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-xyz8
status:
  capacity:
    cpu: "4"
    ephemeral-storage: 101430960Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 15335536Ki
    pods: "110"
    example.com/gpu: 2
```

With this motivating example in mind. We define the following goals and
non-goals of this KEP.

### Goals

* Enable cluster administrators to specify devices advertised by DRA drivers to satisfy
  extended resource requests.

* Enable application operators to use the existing extended resource request in

>   pod spec to request for DRA resources.

Author: fixed

   node that advertises the extended resource in DRA `ResorceSlice` or extended
   resources backed by device plugin. When using DRA devices, it creates a
   special `ResourceClaim` for the pod with the allocation result recording
   which devices were picked. More details on this special `ResourceClaim`
   follow below.  When using extended resources advertised for a node by device
   plugin, the existing resource tracking reserves them.
1. Introduce a field `ExtendedResourceClaimStatus` to pod's `Status`, such that:
    - the kubelet can find the special `ResourceClaim` while looking for claims to prepare
    -  the kubelet can pass the devices to containers in the pod with the extended
   resource requests, based on the container/extended resource to device
   request mapping in the `ExtendedResourceClaimStatus`.

Some quick clarifications around the basic concepts: extended resource backed by
device plugin, extended resource backed by DRA, and dynamic resource.

* extended resource backed by device plugin uses pod's
  spec.containers[].resources.requests to request for resources, it consumes the capacity
  from node's status.capacity. It is of type (string, int64)

> * dynamic resource uses `ResourceClaim` to request for resources, and

Author: fixed.

    requests for foo1.domain/bar and foo2.domain/bar, the allocation of devices
    for each are recorded in DeviceResourceRequestAllocationResult, and just 
    one claim object with allocation `Results` that lists all allocated devices
    is created for the pod.

The special resource claim object lifecycle is managed by the scheduler and
garbage collector.

  * It is *created* in a namespace when there is a pod with extended resource
  request, and the extended resource is advertised by `ResourceSlice` and
  scheduler has fit the pod to a node with the `ResourceSlice`.
  * It is *created* by the scheduler dynamic resource plugin during
    preBind phase. The in-memory one in the assumed cache is created earlier
    during Reserve phase.
  * It is *deleted*
    * either together with the owning pod's deletion.
    * or by the scheduler dynamic resource plugin during unReserve phase.
    * or by the scheduler dynamic resource plugin during PostFilter phase.
  * It is *discovered* by the kubelet via `pod.Status.ExtendedResourceClaimStatus`
  * It is *read* by the kubelet DRA device driver to prepare the devices listed
    therein when preparing to run the pod.

```go

> type DeviceRequest struct {

Author: no, it is not changed. i added it here to show the Name must be a DNS label, and the following section explains how to construct the 'Name' in DeviceRequest.

```go
type DeviceRequest struct {
	// Name can be used to reference this request in a pod.spec.containers[].resources.claims
	// entry and in a constraint of the claim.
	//
	// Must be a DNS label.
	//
	// +required
	Name string `json:"name" protobuf:"bytes,1,name=name"`
}
```

To enable the kubelet to map devices back to the containers which requested them,
the kube-scheduler creates one `DeviceRequest` per extended resource backed by DRA
per container in the pod. The name of the `DeviceRequest` has the form
"container-%d-request-%d", where the first %d is the index of the container in the pod.
The second %d is the index of the extended resource inside the container
resource requests. For example, if the first container in the pod has an
extended resource backed by DRA which is the 3rd such request in the container,

> then the name of the `DeviceRequest` is "container-0-request-2".

Reviewer 2: ```suggestion
then the name of the `DeviceRequest` is "container-0-request-2".

Documenting this naming is merely informational, it is not part of the API. The kubelet must not rely on it. Instead, the `ContainerExtendedResourceRequest` field below specifies the mapping.
```

> Documenting this naming is merely informational, it is not part of the API. The kubelet must not rely on it. Instead, the `ContainerExtendedResourceRequest` field below specifies the mapping.

Author: IPPU allows updating the values of existing resources, need to double check if it allows adding / removing resources.

CPU and memory should be fine with this scheme, as they have their own index in the Requests.

> Documenting this naming is merely informational, it is not part of the API. The kubelet must not rely on it. Instead, the `ContainerExtendedResourceRequest` field below specifies the mapping.

Author: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources  allows 'Resize' only.

### Pod API

A new field `extendedResourceClaimStatus` is added to Pod's status to track
the special resouceclaim object created for the extended resource requests

> in the pod. This is needed for kublet to pass the devices allocated by driver

Author: fixed.

// state of a system, especially if the node that hosts the pod cannot contact the control
// plane.
type PodStatus struct {
        // Status of resource claims.
        // +patchMergeKey=name
        // +patchStrategy=merge,retainKeys
        // +listType=map 
        // +listMapKey=name
        // +featureGate=DynamicResourceAllocation
        // +optional
        ResourceClaimStatuses []PodResourceClaimStatus `json:"resourceClaimStatuses,omitempty" patchStrategy:"merge,retainKeys" patchMergeKey:"name" protobuf:"bytes,15,rep,name=resourceClaimStatuses"`
        // Status of extended resource claim backed by DRA.
        // +featureGate=DynamicResourceAllocation
        // +optional
        ExtendedResourceClaimStatus *PodExtendedResourceClaimStatus `json:"extendedResourceClaimStatus,omitempty" protobuf:"bytes,17,opt,name=extendedResourceClaimStatus"`
}
```

For example, if a pod has requested for foo.domain/bar, and it is
scheduled to run on a node where foo.domain/bar was mapped to devices in a DeviceClass,
then the pod's status is like below:

```yaml
status:
   extendedResourceClaimStatus:

>    - names:

Author: you're right, fixed.

     - container-name
     - foo.domain/bar
     - container-0-request-2
   resourceClaimName: ccc-gpu-57999b9c4c-vpq68-gpu-8s27z
```
where `deviceRequest` name is "container-0-request-2", and container-name is the first container
in the pod, foo.domain/bar is the 3rd extended resource in the container's requests.

Note the validations for extendedResourceClaimStatus are different from the
validations for resourceClaimStatuses.

1. resourceClaimStatuses requires `name` must be DNS label,
   extendedResourceClaimStatus's names' `containerName` and `RequestName` must
   be a DNS label, while the `extendedResourceName` is not a DNS label.
1. resourceClaimStatuses requires `name` must be one of the claim's name in the
   pod spec. extendedResourceClaimStatus requires `containerName` must be one

>    of the container name in the pod spec, and `extendedReourceName` must be one

Reviewer 9: ```suggestion
   of the container name in the pod spec, and `extendedResourceName` must be one
```

   of the extended resource name in that container.

### Scheduling for Extended Resource backed by DRA

A new field `DynamicResources` is added to
[`Resource`](https://github.com/kubernetes/kubernetes/blob/c81431de59a3bf516489317433a165b050322339/pkg/scheduler/framework/types.go#L798),
it works similar to ScalarResources. It is used to keep track of the extended
resources backed by DRA on a node, i.e. those that are advertised by `ResourceSlice`.

```go
type Resource struct {
	MilliCPU         int64
	Memory           int64
	EphemeralStorage int64
	// We store allowedPodNumber (which is Node.Status.Allocatable.Pods().Value())
	// explicitly as int, to avoid conversions and improve performance.
	AllowedPodNumber int
	// ScalarResources
	ScalarResources map[v1.ResourceName]int64

	// NEW!
	// DynamicResources: keep track of extended resources backed by DRA to device classes
	// The map's key is the extended resource name that has at least one device
	// class advertises it.

> 	DynamicResources map[v1.ResourceName][]string

Reviewer 9: "at least one" - but I think we said "exactly one" above, correct? So this can become a `map[v1.ResourceName]string`?

}
```

type `NodeInfo` is used by scheduler to keep track of the information for each
node in memory. Its `Allocatable` field is used to keep track of the allocatable
resources in memory. At the beginning of each scheduling cycle, scheduler takes
a snapshot of all the nodes in the cluster, and updates their corresponding
`NodeInfo`.

For the scheduler with DRA enabled, right after taking the node snapshot, the
scheduler also takes a snapshot of `DeviceClass`, and updates

> `NodeInfo.DynamicResources` if there is extended resource backed by DRA.

Reviewer 9: s/is extended/is an extended/

Author: fixed.

For a node with extended resources from device plugin, its NodeInfo's
Allocatable.ScalarResources is updated with the k8s `Node`'s object.
For a node with extended resources backed by DRA, its NodeInfo's
Allocatable.DynamicResources is updated based on DRA `DeviceClass` objects.

The existing 'noderesources' plugin needs to be modified, such that a pod's
extended resource request is checked against a NodeInfo's ScalarResources if the
node uses device plugin, and checked against a NodeInfo's DynamicResources, if
the request is for extended resources backed by DRA, then 'noderesources' plugin
would pass and leave it to 'dynamicresource' plugin to check if it can be satisfied.

The existing 'dynamicresources' plugin needs to be modified to account for the
extended resource backed by DRA requests.

#### EventsToRegister
This registers all cluster events that might make an unschedulable pod schedulable,
like finishing the allocation of a claim, or resource slice updates.

The existing dynamicresurce plugin has registered all the events needed or

> extended reource backed by DRA, hence no change is needed.

Reviewer 9: ```suggestion
extended resource backed by DRA, hence no change is needed.
```

#### PreFilter
It checks if the pod has any container requests for extended resources backed by DRA.
If not, and no claims in the pod, then the plugin can return early, as there
is nothing to do.

If the pod still needs to be considered by the plugin, then it checks if the
special resource claim for extended resources backed by DRA has been created
before by scheduler, by checking resource claim name having pod name in the
annotation `resource.kubernetes.io/extened-resource-claim: pod-name`.

If found, scheduler would reuse it. If not found, scheduler would create a
special resource claim that has empty spec. The exact spec needs to be decided
during Filter phase, as some node may have device plugin provide the capacity
for the extended resource, some other node may have DRA provide the capacity.

> The requsts in the special resource claim need to vary for each node.

Reviewer 9: ```suggestion
The requests in the special resource claim need to vary for each node.
```

#### Filter

> If a pod has extended resource backed by DRA, and the node does *not* have

Reviewer 9: ```suggestion
If a pod has an extended resource backed by DRA, and the node does *not* have
```

device plugin to provide the capacity for the resource, then the
dynamicresource plugin needs to try allocate the resource by filling in the
special claim's `Spec.Devices.Requests` field.

> One `request` is created per container, and per extended resource backed by DRA

Reviewer 9: so, AIUI, many nodes are evaluated in parallel during Filter, is that correct? And the basic idea here is that you create a "stub" resource claim and then for each node, you create an in-memory spec, that somehow gets written to the API server if and only if that node is the chosen node?

Author: correct, nodes are evaluated in parallel during Filter.

we create a sbut/template claim in PreFilter, and instantiate its Spec during Filter for each node.

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.
-->

End-to-end testing depends on a working resource driver and a container runtime
with CDI support. A [test
driver](https://github.com/kubernetes/kubernetes/tree/master/test/e2e/dra/test-driver)
was developed as part of the overall DRA development effort. We will add tests to
ensure `ExtendedResourceName`s are handled by the scheduler as described in this KEP.

### Graduation Criteria

#### Alpha

- Feature implemented behind a feature flag
- Initial e2e tests completed and enabled

#### Beta

- Gather feedback
- Additional tests are in Testgrid and linked in KEP

#### GA

- 3 examples of vendors making use of the extensions proposed in this KEP

> - Scalability tests that mirror real-world usage as determined by user feedback

Reviewer 9: this will need to move to beta criteria based on https://github.com/kubernetes/enhancements/pull/5242

In general we are raising the bar for beta, so take a look at that and revise these criteria as needed.

Author: done

drivers that they provide to customers.

### Version Skew Strategy

All of the API extensions proposed in this KEP is the optional
`ExtendedResourceName` in `DeviceClass`, and `ExtendedResourceClaimStatus` in
`Pod`. There is no risk for version skew downgrades
because these `DeviceClass` and `Pod` will never have existed in
older clusters.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: DRAExtendedResource
  - Components depending on the feature gate:
    - kube-apiserver
    - kube-scheduler
    - kubelet

###### Does enabling the feature change any default behavior?

No

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Applications that were already deployed and are running will continue to
work. They will continue to work when restarted because the CDI devices that
have been prepared for them won't change across the restart.

The DRA driver itself should also be able to survive a rollback, as there is no

> DRA driver change in this KEP.

Reviewer 9: When the pod is deleted, and the node unprepare resources is called, will it work as expected?

Author: Yes, there is no difference in this respect between a reguar claim, and the special claim backed by DRA. They are handled the same way in kubelet.

###### What happens if we reenable the feature if it was previously rolled back?

The scheduler may lose track of what devices it has allocated to what pods. Any
pods that had previously allocated devices with the feature enabled will need

> to be deleted to ensure they are freed back to their corresponding driver  and

Reviewer 9: ```suggestion
to be deleted to ensure they are freed back to their corresponding driver and
```

  - Details:
-->
Will be considered for beta.

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

<!--
Describe the metrics themselves and the reasons why they weren't added (e.g., cost,
implementation difficulties, etc.).
-->
Will be considered for beta.

### Dependencies

<!--
This section must be completed when targeting beta to a release.
-->

###### Does this feature depend on any specific services running in the cluster?

<!--
Think about both cluster-level services (e.g. metrics-server) as well
as node-level agents (e.g. specific version of CRI). Focus on external or
optional services that are needed. For example, if this feature depends on
a cloud provider API, or upon an external software-defined storage or network
control plane.

For each of these, fill in the following—thinking about running existing user workloads
and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:
-->
No.

### Scalability

> No. The API extensions in this KEP are limited to at most one claim for extended resource backed by DRA per pod.

Reviewer 9: did the question get dropped? or was this for some other question and is duplicated here?

Author: indeed, added back the question.

```

```yaml
apiVersion: resource.k8s.io/v1beta1
kind: ResourceSlice
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2-gpu.coqj92d
spec:
  devices:
  - basic:
    name: gpu-0
  - basic:
    name: gpu-1
  - basic:
    name: gpu-2
  - basic:
    name: gpu-3
  - basic:
    name: gpu-4
  - basic:
    name: gpu-5
  - basic:
    name: gpu-6
  - basic:
    name: gpu-7
  driver: gpu.example.com
  nodeName: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2
```

```yaml
apiVersion: v1
kind: Node
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-zrw2
status:
  capacity:
    cpu: "4"
    ephemeral-storage: 101430960Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 15335536Ki
    pods: "110"
```

```yaml
apiVersion: v1
kind: Node
metadata:
  name: gke-drabeta-n1-standard-4-2xt4-346fe653-xyz8
status:
  capacity:
    cpu: "4"
    ephemeral-storage: 101430960Ki
    hugepages-1Gi: "0"
    hugepages-2Mi: "0"
    memory: 15335536Ki
    pods: "110"
    example.com/gpu: 2
```

With this motivating example in mind. We define the following goals and
non-goals of this KEP.

### Goals

* Introduce the ability to advertise DRA resources as extended resources, and

>   for the scheduler to consider them for allocation.

Author: fixed. thanks.

* Enable application operators to use the existing extended resource request in
  pod spec to request for DRA resources.

* Extended resource support is not added just for easing the transition to DRA
  for the short term. Its ease of use is one big advantage to keep it remaining
  useful for the long term.

* Device plugin API must not change. The existing device plugin drivers must
  continue working without change.

* DRA driver API must not change. Core kubernetes (kube-scheduler, kubelet) is
  preferred over DRA driver for any change needed to support the feature.

### Non-Goals

* Minimize kubelet or kube-scheduler changes. The feature requires necessary
  changes in both scheduling and actuation.

* Keep advertising `node.status.Capacity` for extended resources backed by DRA.

>   It is used for extended resources backed by device plugin only.

Author: the idea here is: it is *NOT* a goal to advertise extended resources backed by DRA in `node.status.Capacity`

that is: `node.status.Capacity` is only for extended resources backed by device plugin, *NOT* for extended resources backed by DRA

>   It is used for extended resources backed by device plugin only.

Reviewer 2: > it is NOT a goal to advertise extended resources backed by DRA in node.status.Capacity

That this is not a goal does not mean that we won't do it. It could be a side-effect of the implementation.

Better use "Goals" for the hard promises that you are making and "Non-goals" for things that a reader might expect, but can't be promised. It's about managing expectations, not a plan of record of things that definitely won't be done.

>   It is used for extended resources backed by device plugin only.

Author: @klueska was expecting to have 'extended resource backed by DRA' updte node.status.Capacity last time when he reviewed this (a few weeks ago), and he suggested we made it clear that this KEP is not planning to doing that in the Non-goals section.

what is your suggestion? remove it from Non-goals?

>   It is used for extended resources backed by device plugin only.

Reviewer 2: My suggestion is to remove "Keep advertising `node.status.Capacity` for extended resources backed by DRA. ..." from non-goals and to add "Keep advertising only extended resources backed by Device Plugins in `node.status.Capacity`." to goals.

>   It is used for extended resources backed by device plugin only.

Author: Got it. done. PTAL.

## Proposal

The basic idea is the following:

1. Introduce `extended resource backed by DRA`. It is like the current extended
   resource backed by device plugin, in that, it has a string name, and a
   discrete countable quantity. Its capacity is provided through DRA
   `ResourceSlice`, its consumption is specified through pod's extended
   resource request.
1. Introduce a field `ExtendedResourceName` to `DeviceClass` to allow cluster
   administrators to advertise certain class of devices as extended resource.

> 1. Introduce a special `ResourceClaim` object to keep track of device allocations

Reviewer 2: Okay, let's call it "special". But then let's also immediately clarify how it is "special".

```suggestion
1. Introduce a special `ResourceClaim` object to keep track of device allocations. It is special only in the sense that it is created by the scheduler. No semantic changes are needed in the API for it.
```

   for all extended resource requests backed by DRA for a pod. kube-scheduler
   uses DRA scheduling algorithm to fit pod's extended resource request to a
   node that advertises the extended resource in DRA `ResorceSlice` or extended
   resources backed by device plugin. When using DRA devices, it creates a
   special `ResourceClaim` for the pod with the allocation result recording
   which devices were picked. More details on this special `ResourceClaim`
   follow below.  When using extended resources advertised for a node by device
   plugin, the existing resource tracking reserves them.
1. kubelet asks DRA driver to prepare devices in the special `ResourceClaim`,

>    and pass the devices to containers in a pod with the extended resource requests.

Author: added quick summary on how kublet uses extendedResourceClaimStatus in pod.status, and the special resource claim to ask DRA driver to prepare the devices, and pass those devices to the proper containers correctly

  named extended resource backed by device plugin at the same time. This implies
  that either the resource is advertised through `ResourceSlice`,
  or `Node`'s status.capacity.

## Design Details

### Device Class API
The extended resource name to DRA device mapping can be specified at
`DeviceClassSpec`. The same extended resource name can be given to different
device classes, and one device class can have at most one extended resource name.

Cluster administrator is soly responsible for creating device classes, and the
mapping between the class of devices and the extended resource name.
`DeviceClass` is cluster scoped, application developers and operators cannot change it.

The mapping of DRA devices and extended resources is stored in k8s data store
(e.g. etcd). It is created after cluster creation, before deployment of the
application that uses the devices.

```go
// DeviceClassSpec is used in a [DeviceClass] to define what can be allocated
// and how to configure it.
type DeviceClassSpec struct {
	// Each selector must be satisfied by a device which is claimed via this class.
	//
	// +optional
	// +listType=atomic

> 	Selectors []DeviceSelector `json:"selectors,omitempty" protobuf:"bytes,1,opt,name=selectors"`

Author: truncated.

	// Config defines configuration parameters that apply to each device that is claimed via this class.
	// Some classses may potentially be satisfied by multiple drivers, so each instance of a vendor
	// configuration applies to exactly one driver.
	//
	// They are passed to the driver, but are not considered while allocating the claim.
	//
	// +optional
	// +listType=atomic
	Config []DeviceClassConfiguration `json:"config,omitempty" protobuf:"bytes,2,opt,name=config"`

	// ExtendedResourceName is the extended resource name the device class is advertised as.
	// All devices matched by the device class can be used to satisfy the extended resource requests in pod's spec.
	//
	// +optional
	ExtendedResourceName *string `json:"extendedResourceName,omitempty" protobuf:"bytes,4,opt,name=extendedResourceName"`
}
```

### Resource Claim API

> A special resource claim object is created to keep track of device allocations for

Author: It is special, probably not very special (in that the claim content is mostly the same as today), but the way it is constructed (with one device request for each extended resource request in each container) is different from the regular claim.

> A special resource claim object is created to keep track of device allocations for

Reviewer 2: Okay (https://github.com/kubernetes/enhancements/pull/5136#discussion_r2073518326).

extended resource. The resource claim object has following properties:

  * It is namespace scoped, like other resource claim objects.
  * It is owned by a pod, like other resource claim objects.
  * It has `spec` of devices.requests, with each request name being a hash of the

>     container name and the extended resource backed by DRA name inside the container.

Author: updated to make it clearer. also replaced hash with encoding.

  * Its `status.allocation.devices` and `status.allocation.reservedFor` are
    used.

>   * It does not have annotation `resource.kubernetes.io/pod-claim-name:` as

Author: Yes, added the new annotation:
resource.kubernetes.io/owned-by-scheduler: true

sheduler can use it to find the special claim, and either reuse it (if its allocation is still valid), or deallocate/delete it during postFilter phase, such that a new special resource claim can be created anew.

    it is created for the extended resource request in a pod spec, not for a
    claim in the pod spec.
  * At most one such claim object is created per pod. For example, if the pod
    requests for foo1.domain/bar and foo2.domain/bar, the allocation of devices
    for each are recorded in DeviceResourceRequestAllocationResult, and

>     one claim object with two `Results` is created for the pod.

Author: reworded.

The special resource claim object lifecycle is managed by the scheduler and
garbage collector.

  * It is *created* in a namespace when there is a pod with extended resource
  request, and the extended resource is advertised by `ResourceSlice` and
  scheduler has fit the pod to a node with the `ResourceSlice`.
  * It is *created* by the scheduler dynamic resource plugin during
    preBind phase. The in-memory one in the assumed cache is created earlier

>     during Reserve phase.

Author: agreed, we will use the cleaner way, adding pod.status.extendedResourceClaimName field

  * It is *deleted*
    * either together with the owning pod's deletion.
    * or by the scheduler dynamic resource plugin during unReserve phase.

>   * It is *read* by the kubelet DRA device driver to prepare the devices listed

Author: added the following:

It is *discovered* by the kubelet via `pod.Status.ExtendedResourceClaimStatus`

how kubelet looks up the name is provided below

    therein when preparing to run the pod.

```go
type DeviceRequest struct {
	// Name can be used to reference this request in a pod.spec.containers[].resources.claims
	// entry and in a constraint of the claim.
	//
	// Must be a DNS label.
	//
	// +required
	Name string `json:"name" protobuf:"bytes,1,name=name"`
}
```

As shown above, device request must have a name, and it must be a DNS label.
kube-scheduler creates one `DeviceRequest` per extended resource backed by DRA
per container in the pod. The name of the `DeviceRequest` is determined by the
container name, and the extended resource backed by DRA name. It is in the
form of "c%d-e%d", where the first %d is the index of the container in the pod.
The second %d is the index of the extended resource inside the container
resource requests. For example, if the first container in the pod has an
extended resource backed by DRA which is the 3rd such request in the container,
then the name of the `DeviceRequest` is "c0-e2".

### Pod API

> A new field `extendedResourceClaimStatus` is added to Pod's status to track

Reviewer 2: Looks like you arrived at a similar conclusion as I did above... good! :smile: 

Let's make sure that this is called out above, otherwise readers will be kept scratching their head about how this will work for too long.

Author: Yes, added `extendedResourceClaimStatus` earlier in the doc such that readers can get the basic idea.

the special resouceclaim object created for the extended resource requests
in the pod. This is needed for kublet to pass the devices allocated by driver
to the containers in the pod.

```go
// PodExtendedResourceClaimStatus is stored in the PodStatus for each extended
// resource requests backed by DRA. It stores the generated name for 
// the corresponding special ResourceClaim created by scheduler.
type PodExtendedResourceClaimStatus struct {
        // Names identifies the mapping of <container, extended resource backed by DRA> to  device request.
        // +patchMergeKey=requestName
        // +patchStrategy=merge,retainKeys
        // +listType=map
        // +listMapKey=requestName
        // +featureGate=DynamicResourceAllocation

>         Names []ContainerExtendedResourceRequest `json:"names" patchStrategy:"merge,retainKeys" patchMergeKey:"requestName" protobuf:"bytes,1,rep,name=names"`

Reviewer 2: I am not sure what this field is needed for.

If we use the `container-%d-request-%d` naming specified above, then the kubelet can parse that to determine the container.

If the purpose of this field is to avoid such string parsing, then it needs to be called out above that the request names are arbitrary and that the `container-%d-request-%d` scheme is not required. This is probably cleaner.

If we keep this struct as-is for that reason, then make this a `+listType=atomic`. It's simpler and more efficient (no need to track individual owners in managed fields when there is always only exactly one owner).

Reviewer 2: This wasn't clear above. Let's clarify as suggested in https://github.com/kubernetes/enhancements/pull/5136/files#r2073542772

        // created for extended resource requests backed by DRA.
        RequestName string `json:"requestName" protobuf:"bytes,3,name=requestName"`
}

// PodStatus represents information about the status of a pod. Status may trail the actual
// state of a system, especially if the node that hosts the pod cannot contact the control
// plane.
type PodStatus struct {
        // Status of resource claims.
        // +patchMergeKey=name
        // +patchStrategy=merge,retainKeys
        // +listType=map 
        // +listMapKey=name
        // +featureGate=DynamicResourceAllocation
        // +optional
        ResourceClaimStatuses []PodResourceClaimStatus `json:"resourceClaimStatuses,omitempty" patchStrategy:"merge,retainKeys" patchMergeKey:"name" protobuf:"bytes,15,rep,name=resourceClaimStatuses"`
        // Status of extended resource claim backed by DRA.
        // +featureGate=DynamicResourceAllocation
        // +optional
        ExtendedResourceClaimStatus *PodExtendedResourceClaimStatus `json:"extendedResourceClaimStatus,omitempty" protobuf:"bytes,17,opt,name=extendedResourceClaimStatus"`
}
```

For example, if a pod has requested for foo.domain/bar, and it is

> scheduled to run on a node that has advertised foo.domain/bar in `ResourceSlice`,

Reviewer 2: ```suggestion
scheduled to run on a node that where foo.domain/bar was mapped to devices in a DeviceClass,
```

node uses device plugin, and checked against a NodeInfo's DynamicResources, if
the request is for extended resources backed by DRA, then 'noderesources' plugin
would pass and leave it to 'dynamicresource' plugin to check if it can be satisfied.

The existing 'dynamicresources' plugin needs to be modified to account for the
extended resource backed by DRA requests.

#### EventsToRegister
This registers all cluster events that might make an unschedulable pod schedulable,
like finishing the allocation of a claim, or resource slice updates.

The existing dynamicresurce plugin has registered all the events needed or
extended reource backed by DRA, hence no change is needed.

#### PreFilter
It checks if the pod has any container requests for extended resources backed by DRA.
If not, and no claims in the pod, then the plugin can return early, as there
is nothing to do.

If the pod still needs to be considered by the plugin, then it creates a
special resource claim that has empty spec. The exact spec needs to be decided
during Filter phase, as some node may have device plugin provide the capacity
for the extended resource, some other node may have DRA provide the capacity.

> The requsts in the special resource claim need to vary for each node.

Reviewer 2: In PreFilter we also need to pick up a stale ResourceClaim that is not yet linked to in the pod status.

Author: good point. added that PreFilter would reuse the special ResourceClaim if it was created before. it would create a new one only when no such claim exists.

#### Filter
If a pod has extended resource backed by DRA, and the node does *not* have
device plugin to provide the capacity for the resource, then the
dynamicresource plugin needs to try allocate the resource by filling in the
special claim's `Spec.Devices.Requests` field.

One `request` is created per container, and per extended resource backed by DRA
in the container. The `DeviceClass` in the request is randomly picked if there

> are multiple device classes advertising the extended resource. The `Name` of the

Reviewer 2: Picking one class locks selection of devices to that class, which is probably not what admins expected.

We either have to say "exactly one device class may be mapped to one extended resource name" and make it an error if more then one is found, or support selecting devices from any of the classes which where mapped to the same name.

Author: Let's start with more restrictive, and simpler one. if needed, we can relax it in the future. reworded the doc.

"exactly one device class may be mapped to one extended resource name" and make it an error if more then one is found"

Reviewer 2: We could support multiple classes by using a prioritized list. Then we can support multiple classes, up to the maximum size of the prioritized list. We could add that later, but then would need another release cycle where we keep the feature at alpha according to the upcoming "feature must be complete before graduation to beta" rule.

Perhaps we should try to do it in 1.34 after all?

cc @klueska

Author: I'm open to the suggestion, esp if you think it is common for cluster admins to use multiple classes for one extended resource.

request is determined by the container name and the extended resource name, in
case the name conflicts with an existing device request, the next choice is

> deterministically picked until there is no conflict.

Reviewer 2: What conflict? All request names in the ResourceClaim created for the extended resources are under control of the scheduler.

Author: there should not be conflict, removed it.

The allocator needs to be modified to allow for the special resource claim for
extended resource backed by DRA, which could vary by node. The `Allocate`

> method takes the claim as a parameter, in adddition to node parameter. The

Reviewer 2: Why can this claim not be another entry in `claimsToAllocate`?

https://github.com/kubernetes/kubernetes/blob/680ea07dbb2c6050d13b93660fa4d27d2d28d6eb/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L72

Author: In the prototype implementation, I made all changes (there are not many) in the scheduler plugin to add the special resourceclaim to the call `Allocate`. It seems `claimsToAllocate` does *NOT* change across nodes, so it is reasonable to leave it in `NewAllocator`.

Author: i see what you mean. claimsToAllocate has one special resource claim template (it does not have spec.Requests), which is created during the PreFilter, the template is instantiated (spec.Requests are filled in based on each node) during Filter, then inside Allocator, if the one being processed is the claim template, then it is replaced with the instantiated claim passed in through the Allocate()

https://github.com/kubernetes/kubernetes/pull/130653 has the prototype code.

Reviewer 2: This looks complicated to me.

Reviewer 3: > @yliaog: 
> i see what you mean. claimsToAllocate has one special resource claim template (it does not have spec.Requests), which is created during the PreFilter, the template is instantiated (spec.Requests are filled in based on each node) during Filter, then inside Allocator, if the one being processed is the claim template, then it is replaced with the instantiated claim passed in through the Allocate()
>
> kubernetes/kubernetes#130653 has the prototype code.

> @pohly:
> This looks complicated to me.

I agree with @pohly. The current approach in the POC impl looks like it potentially breaks all consumers of `claimsToAllocate` unless they adapt to understand and handle the unusual not-yet-persisted claim with requests that differ per node ... that isn't great in terms of complexity or performance or principle of least surprise.

If we can encapsulate the per-node nature of the extended resource claim inside state / allocator, ideally in a way that minimizes per-node allocs, that would be ideal

Reviewer 5: This is why it would be better to have one "optional" resource claim which may, but does not have to be allocated. It does not have to be created in api-server upfront (we had long discussion about it already), but could be at least created in-memory in preFilter.

It could behave as a regular resource claim, but filter would not discard a node when it was not allocatable. The fit plugin however would discard a node if there is no extended nor dynamic resource.

Finally, the claim would be on the `claimsToAllocate` list only if it was allocatable for the selected node. I may be still missing something, as I suspect this fairly simple approach was discarded for some reason, but it would be good to know what the blocker is.

The "optional" nature of the resource claim would not have to be exposed in the API and contained in the scheduling logic, so the overall goal to not change the API would be preserved.

algorithm uses the passed in special claim whenever it processes the last claim
in the claims slice, which is an instantiation of the special claim template
created during the preFilter phase.

If there is an allocation for a node, the allocation, and the claim are
recorded in cyclestate.

#### Score
If two nodes can fit the pod, one node has installed device plugin, the other
has installed DRA, score the DRA node higher. This is not going to be

> considered in ALPHA.

Reviewer 2: I'm not sure whether this scoring is useful.

If both device plugin and DRA driver provide access to the same kind of hardware, then one method is not necessarily "better" than the other.

Or is the intent to automatically migrate workloads off device plugins? Then state that as the motivation here.

Author: my intention is to automatically migrate workloads off device plugins, but i agree we should start simple, and iterate later in the future. hence, removed the whole scorer section for now.

   pass the proper allocated device IDs to the proper container.

### Test Plan

<!--
[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
-->

[x] I/we understand the owners of the involved components may require updates to
existing tests to make this code solid enough prior to committing the changes necessary
to implement this enhancement.

##### Prerequisite testing updates

##### Unit tests

<!--
Additionally, for Alpha try to enumerate the core package you will be touching
to implement this enhancement and provide the current unit coverage for those
in the form of:
- <package>: <date> - <current test coverage>
The data can be easily read from:
https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-coverage-unit

This can inform certain test coverage improvements that we want to do before
extending the production code to implement this enhancement.
- `<package>`: `<date>` - `<test coverage>`
-->

Start of v1.33 development cycle (v1.33.0-alpha.x-xxx-xxxxxxxxxxxx):

- `k8s.io/dynamic-resource-allocation/cel`: ??.?%
- `k8s.io/dynamic-resource-allocation/structured`: ??.?%
- `k8s.io/kubernetes/pkg/controller/resourceclaim`: ??.?%

> - `k8s.io/kubernetes/pkg/scheduler/framework/plugins/dynamicresources`: ??.?%

Reviewer 2: Please fill this in for components that you need to modify. Can wait for 1.33.0.

Author: done
